{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Experiment with text generation for BMO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import getpass\n",
    "import json\n",
    "import random\n",
    "import rdflib\n",
    "import re\n",
    "import requests\n",
    "import wikipedia\n",
    "import urllib\n",
    "\n",
    "from bs4 import BeautifulSoup\n",
    "from collections import Counter, defaultdict\n",
    "\n",
    "import numpy as np\n",
    "import networkx as nx\n",
    "\n",
    "import spacy\n",
    "\n",
    "from sklearn.model_selection import train_test_split\n",
    "\n",
    "from bmo_tools.ontologies import subontology_from_term\n",
    "\n",
    "from bluegraph import PandasPGFrame\n",
    "from bluegraph.backends.networkx import pgframe_to_networkx\n",
    "\n",
    "from kgforge.core.forge import KnowledgeGraphForge\n",
    "from kgforge.core.resource import Resource\n",
    "from kgforge.specializations.resources import Dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "········\n"
     ]
    }
   ],
   "source": [
    "TOKEN = getpass.getpass()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "ENDPOINT = \"https://bbp.epfl.ch/nexus/v1\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "forge = KnowledgeGraphForge(\n",
    "    \"https://raw.githubusercontent.com/BlueBrain/nexus-forge/master/examples/notebooks/use-cases/prod-forge-nexus.yml\",\n",
    "    endpoint=ENDPOINT,\n",
    "    token=TOKEN,\n",
    "    bucket=\"bbp/kg-to-text\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def is_nan(el):\n",
    "    try:\n",
    "        if np.isnan(el):\n",
    "            return True\n",
    "    except Exception as e:\n",
    "        pass\n",
    "    return False\n",
    "\n",
    "def search_wiki(query):\n",
    "    url = \"https://en.wikipedia.org/w/api.php\"\n",
    "    \n",
    "    params = {\n",
    "        \"action\": \"opensearch\",\n",
    "        \"namespace\": \"0\",\n",
    "        \"limit\": \"5\",\n",
    "        \"format\": \"json\",\n",
    "        \"search\": query\n",
    "    }\n",
    "\n",
    "    result = requests.get(url, params=params).json()\n",
    "    _, titles, _, uris = result\n",
    "    if len(titles) > 0:\n",
    "        return titles[0], uris[0]\n",
    "    \n",
    "def query_dbpedia(wiki_url):\n",
    "    suffix = wiki_url.split(\"/\")[-1]\n",
    "    dbpedia_id = f\"http://dbpedia.org/data/{suffix}.json\"\n",
    "\n",
    "    res = requests.get(dbpedia_id).json()\n",
    "    incoming_links = []\n",
    "    for k, v in res.items():\n",
    "        if k != f\"http://dbpedia.org/resource/{suffix}\":\n",
    "            incoming_links.append(k)\n",
    "    data = res[f\"http://dbpedia.org/resource/{suffix}\"]\n",
    "\n",
    "    return data, incoming_links\n",
    "\n",
    "\n",
    "def search_dbpedia(query):\n",
    "    url = \"https://lookup.dbpedia.org/api/search\"\n",
    "    \n",
    "    params = {\n",
    "        \"format\": \"json\",\n",
    "        \"query\": query\n",
    "    }\n",
    "    \n",
    "    result = requests.get(url, params=params).json()\n",
    "    results = []\n",
    "    for el in result[\"docs\"]:\n",
    "        record = {}\n",
    "        record[\"url\"] = el['resource'][0]\n",
    "        record[\"redirecredLabels\"] = set([\n",
    "            l.replace(\"<B>\", \"\").replace(\"</B>\", \"\")\n",
    "            for l in el.get('redirectlabel', [])\n",
    "        ])\n",
    "        record[\"label\"] = [\n",
    "            l.replace(\"<B>\", \"\").replace(\"</B>\", \"\")\n",
    "            for l in el.get('label', [])\n",
    "        ]\n",
    "        record[\"type\"] = el.get(\"type\", None)\n",
    "        results.append(record)\n",
    "    return results\n",
    "\n",
    "\n",
    "def dbpedia_uri_to_wiki(uri):\n",
    "    suffix = uri.split(\"/\")[-1]\n",
    "    return f\"https://en.wikipedia.org/wiki/{suffix}\"\n",
    "    return wikipedia_uri\n",
    "\n",
    "\n",
    "def wikipedia_uri_to_db(uri):\n",
    "    suffix = wiki_url.split(\"/\")[-1]\n",
    "    return f\"http://dbpedia.org/resource/{suffix}\"\n",
    "\n",
    "\n",
    "def extract_wiki_abstract(wiki_url):\n",
    "    suffix = wiki_url.split(\"/\")[-1]\n",
    "    dbpedia_id = f\"http://dbpedia.org/data/{suffix}.json\"\n",
    "\n",
    "    res = requests.get(dbpedia_id)\n",
    "    data = res.json()[f\"http://dbpedia.org/resource/{suffix}\"]\n",
    "\n",
    "    vals = [\n",
    "        el[\"value\"]\n",
    "        for el in data.get(\"http://dbpedia.org/ontology/abstract\", [])\n",
    "        if el[\"lang\"] == \"en\"\n",
    "    ]\n",
    "    return vals[0] if len(vals) > 0 else None\n",
    "\n",
    "\n",
    "def extract_full_wiki_text(wiki_url):\n",
    "    \"\"\"Adapted from https://levelup.gitconnected.com/two-simple-ways-to-scrape-text-from-wikipedia-in-python-9ce07426579b.\"\"\"\n",
    "    html_source = requests.get(wiki_url).text\n",
    "    soup = BeautifulSoup(html_source,'lxml')\n",
    "\n",
    "    # Extract the plain text content from paragraphs\n",
    "    paras = []\n",
    "    for paragraph in soup.find_all('p'):\n",
    "        paras.append(str(paragraph.text))\n",
    "\n",
    "    # Interleave paragraphs & headers\n",
    "    text = ' '.join(paras)\n",
    "\n",
    "    # Drop footnote superscripts in brackets\n",
    "    text = re.sub(r\"\\[.*?\\]+\", '', text)\n",
    "\n",
    "    # Replace '\\n' (a new line) with '' and end the string at $1000.\n",
    "    text = text.replace('\\n', '')[:-11]\n",
    "    return text\n",
    "\n",
    "\n",
    "def annotate_text(nlp, text):\n",
    "    url = \"https://api.dbpedia-spotlight.org/en/annotate\"\n",
    "    params={\n",
    "        \"text\": nlp.make_doc(text)\n",
    "    }\n",
    "    params = urllib.parse.urlencode(params, quote_via=urllib.parse.quote)\n",
    "    result = requests.get(url, params=params, headers={\"accept\": \"application/json\"})\n",
    "    try:\n",
    "        json_data = result.json()\n",
    "    except:\n",
    "        json_data = None\n",
    "\n",
    "    return json_data\n",
    "\n",
    "\n",
    "def text_to_batches(text, max_batch_length):\n",
    "    text_length = len(text)\n",
    "    if text_length <= max_batch_length:\n",
    "        yield text\n",
    "    else:\n",
    "        start = 0\n",
    "        end = 0\n",
    "        while True:\n",
    "            if start + max_batch_length > text_length or end == -1:\n",
    "                break\n",
    "            end = text.rfind(\" \", start, start + max_batch_length + 1)\n",
    "            yield text[start:end]\n",
    "            start = end + 1\n",
    "        yield text[start:]\n",
    "        \n",
    "\n",
    "def wiki_url_from_wikidata(wikidata_id, lang):\n",
    "    result =requests.get(\n",
    "        \"https://www.wikidata.org/w/api.php\",\n",
    "        params={\n",
    "            \"action\": \"wbgetentities\",\n",
    "            \"format\": \"json\",\n",
    "            \"props\": \"sitelinks\",\n",
    "            \"ids\": wikidata_id,\n",
    "            \"sitefilter\": lang + \"wiki\"\n",
    "        })\n",
    "    entities = result.json().get(\"entities\")\n",
    "    if entities:\n",
    "        entity = entities.get(wikidata_id)\n",
    "        record = entity[\"sitelinks\"].get(f\"{lang}wiki\")\n",
    "        if record:\n",
    "            title = record[\"title\"]\n",
    "            return f\"https://en.wikipedia.org/wiki/{title.replace(' ', '_')}\"\n",
    "\n",
    "\n",
    "def get_id_by_label(graph, label):\n",
    "    uri = None\n",
    "    for s in graph.subjects(\n",
    "            rdflib.RDFS.label,\n",
    "            rdflib.Literal(label, lang=\"en\")):\n",
    "        uri = s\n",
    "        break\n",
    "    if uri is None:\n",
    "        for s in graph.subjects(\n",
    "                rdflib.RDFS.label,\n",
    "                rdflib.Literal(label, datatype=rdflib.XSD.string)):\n",
    "            uri = s\n",
    "            break\n",
    "    if uri is None:\n",
    "        for s in graph.subjects(\n",
    "                rdflib.RDFS.label,\n",
    "                rdflib.Literal(label)):\n",
    "            uri = s\n",
    "            break\n",
    "    return s\n",
    "\n",
    "\n",
    "def add_annotation(record, label, text, local_start, local_end):\n",
    "    span = {}\n",
    "\n",
    "    span[\"start\"] = local_start + len(record[\"text\"]) + 1\n",
    "    span[\"end\"] = local_end + len(record[\"text\"]) + 1\n",
    "    record[\"text\"] += \" \" + text\n",
    "\n",
    "    span[\"text\"] = label\n",
    "    span[\"label\"] = str(get_id_by_label(rdf_graph, label))\n",
    "    \n",
    "    record[\"spans\"].append(span)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1. Generate simple sentences from BMO\n",
    "\n",
    "Here we generate simple sentences from triples of the ontology:\n",
    "- definition\n",
    "- synonyms\n",
    "- subClassOf relationships"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "frame = PandasPGFrame.from_ontology(\n",
    "        filepath=\"../../ontologies/bbp/bmo.ttl\",\n",
    "        format=\"turtle\", remove_prop_uris=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<Graph identifier=N90fdd4275711483ea73b4a87e92cf682 (<class 'rdflib.graph.Graph'>)>"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rdf_graph = rdflib.Graph()\n",
    "rdf_graph.parse(\"../../ontologies/bbp/bmo.ttl\", format=\"turtle\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "BMO_URI = \"https://bbp.epfl.ch/ontologies/core/bmo/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "graph = pgframe_to_networkx(frame)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "COMMON_EDGE_ALIASES = {\n",
    "    \"IS_SUBCLASS_OF\": \"is a\",\n",
    "    \"isPartOf\": \"is a part of\",\n",
    "    \"hasParameter\": \"has parameter\",\n",
    "    \"modelOf\": \"is a model of\",\n",
    "    \"hasFeature\": \"has feature\",\n",
    "    \"hasType\": \"has type\"\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "graph = pgframe_to_networkx(frame)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "synthetic_text = []"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Node:  EModel Hyperparameter\n",
      "\t [D] EModel Hyperparameter is A parameter used to represent a specific value used in the settings of the EModel building pipeline..\n",
      "\t [R] EModel Hyperparameter is a Hyperparameter.\n",
      "\n",
      "Node:  Hyperparameter\n",
      "\t [R] Hyperparameter is a Entity.\n",
      "\n",
      "Node:  EModel Parameter Optimization Workflow\n",
      "\t [R] EModel Parameter Optimization Workflow is a Parameter Optimization Workflow.\n",
      "\t [R] EModel Parameter Optimization Workflow is a part of EModel Building Workflow.\n",
      "\t [R] EModel Parameter Optimization Workflow generates Ion Channel Density.\n",
      "\t [R] EModel Parameter Optimization Workflow generates Ion Dynamics Feature.\n",
      "\t [R] EModel Parameter Optimization Workflow has parameter EModel Hyperparameter.\n",
      "\t [R] EModel Parameter Optimization Workflow hasTarget Neuron Electrophysiological Feature.\n",
      "\t [R] EModel Parameter Optimization Workflow uses Ion Channel Density Constraint.\n",
      "\t [R] EModel Parameter Optimization Workflow uses Model Ionic Mechanism.\n",
      "\t [R] EModel Parameter Optimization Workflow uses Neuron Morphology.\n",
      "\n",
      "Node:  Parameter Optimization Workflow\n",
      "\t [R] Parameter Optimization Workflow is a Workflow.\n",
      "\t [R] Parameter Optimization Workflow is a part of Model Building Workflow.\n",
      "\t [R] Parameter Optimization Workflow generates Model Brain Parameter.\n",
      "\t [R] Parameter Optimization Workflow has parameter Model Brain Parameter.\n",
      "\t [R] Parameter Optimization Workflow hasTarget Brain Component Feature.\n",
      "\t [R] Parameter Optimization Workflow uses Dataset.\n",
      "\t [R] Parameter Optimization Workflow uses Model Brain Component.\n",
      "\n",
      "Node:  EModel Building Workflow\n",
      "\t [R] EModel Building Workflow is a Neuron Building Workflow.\n",
      "\t [R] EModel Building Workflow generates EModel.\n",
      "\t [R] EModel Building Workflow about Neuron Electrical Type.\n",
      "\t [R] EModel Building Workflow hasTarget Neuron Electrophysiological Feature.\n",
      "\t [R] EModel Building Workflow qualifiedAssociation Association.\n",
      "\t [R] EModel Building Workflow uses Ion Channel Density.\n",
      "\t [R] EModel Building Workflow uses Ion Dynamics Feature.\n",
      "\t [R] EModel Building Workflow uses Model Ionic Mechanism.\n",
      "\t [R] EModel Building Workflow uses Neuron Morphology.\n",
      "\n",
      "Node:  Ion Channel Density\n",
      "\t [D] Ion Channel Density is Density of ion channels in a subcellular component of a neuron..\n",
      "\t [R] Ion Channel Density is a EModel Parameter.\n",
      "\t [R] Ion Channel Density is a Neuron Part Feature.\n",
      "\t [R] Ion Channel Density compartment Neuron Part.\n",
      "\t [R] Ion Channel Density brainRegion Brain Region.\n",
      "\t [R] Ion Channel Density ionChannel Ion Channel.\n",
      "\n",
      "Node:  Ion Dynamics Feature\n",
      "\t [D] Ion Dynamics Feature is Kinetics of the conductance mechanism of an ion..\n",
      "\t [R] Ion Dynamics Feature is a EModel Parameter.\n",
      "\t [R] Ion Dynamics Feature is a Neuron Part Feature.\n",
      "\t [R] Ion Dynamics Feature compartment Neuron Part.\n",
      "\t [R] Ion Dynamics Feature ion Ion.\n",
      "\n",
      "Node:  Neuron Electrophysiological Feature\n",
      "\t [D] Neuron Electrophysiological Feature is An electrical property of a neuron that characterizes a cell's ability to control transmembrane ion flux to enable the generation of action potentials (APs)..\n",
      "\t [R] Neuron Electrophysiological Feature is a Neuron Feature.\n",
      "\t [R] Neuron Electrophysiological Feature stimulusType Electrical Stimulus.\n",
      "\n",
      "Node:  Ion Channel Density Constraint\n",
      "\t [R] Ion Channel Density Constraint is a EModel Parameter Constraint.\n",
      "\t [R] Ion Channel Density Constraint constraints Ion Channel Density.\n",
      "\n",
      "Node:  Model Ionic Mechanism\n",
      "\t [R] Model Ionic Mechanism is a Model Brain Activity.\n",
      "\t [R] Model Ionic Mechanism is a model of Ionic Mechanism.\n",
      "\n",
      "Node:  Neuron Morphology\n",
      "\t [R] Neuron Morphology is a Dataset.\n",
      "\t [R] Neuron Morphology hasMType Neuron Morphological Type.\n",
      "\n",
      "Node:  Synaptic Connection Feature\n",
      "\t [R] Synaptic Connection Feature is a Brain Component Feature.\n",
      "\n",
      "Node:  Brain Component Feature\n",
      "\t [D] Brain Component Feature is A measurable property or feature of a biological or model brain component..\n",
      "\t [R] Brain Component Feature is a Entity.\n",
      "\n",
      "Node:  Ca_HVA2\n",
      "\t [D] Ca_HVA2 is High-voltage-gaterd calcium current.\n",
      "\t [R] Ca_HVA2 is a Ionic Mechanism.\n",
      "\t [R] Ca_HVA2 ion Ca.\n",
      "\n",
      "Node:  Ionic Mechanism\n",
      "\t [D] Ionic Mechanism is A mechanism underlying the dynamics of specific ion current type..\n",
      "\t [R] Ionic Mechanism is a Biological Brain Activity.\n",
      "\t [R] Ionic Mechanism associatedIon Ion.\n",
      "\n",
      "Node:  Ca\n",
      "\t [S] Ca is also know as Calcium.\n",
      "\t [S] Ca is referred to as Calcium.\n",
      "\t [R] Ca is a Ion.\n",
      "\n",
      "Node:  Query\n",
      "\t [R] Query is a Entity.\n",
      "\n",
      "Node:  Entity\n",
      "\t [D] Entity is A Brain Modeling Entity is any entity (biological or modeled) that can be used or generated in the process of digitally reconstructing brain tissue.\n",
      "\n",
      "Node:  Model Mitochondrion Metabolism\n",
      "\t [R] Model Mitochondrion Metabolism is a Model Subcellular Part Metabolism.\n",
      "\t [R] Model Mitochondrion Metabolism is a model of Mitochondrion.\n",
      "\n",
      "Node:  Model Subcellular Part Metabolism\n",
      "\t [R] Model Subcellular Part Metabolism is a Model Subcellular Part.\n",
      "\t [R] Model Subcellular Part Metabolism is a part of Model Glia Metabolism.\n",
      "\t [R] Model Subcellular Part Metabolism is a part of Model Neuron Metabolism.\n",
      "\t [R] Model Subcellular Part Metabolism is a model of Subcellular Part.\n",
      "\n",
      "Node:  Mitochondrion\n",
      "\t [R] Mitochondrion is a Organelle.\n",
      "\n",
      "Node:  EModel Validation Threshold\n",
      "\t [D] EModel Validation Threshold is Threshold under which each score of the EModel has to be for it to pass validation (used by the default validation function)..\n",
      "\t [R] EModel Validation Threshold is a EModel Hyperparameter.\n",
      "\n",
      "Node:  Model kv11\n",
      "\t [R] Model kv11 is a Model Ion Channel.\n",
      "\t [R] Model kv11 is a model of Kv11.\n",
      "\n",
      "Node:  Model Ion Channel\n",
      "\t [R] Model Ion Channel is a Model Brain Component.\n",
      "\t [R] Model Ion Channel is a model of Ion Channel.\n",
      "\n",
      "Node:  Kv11\n",
      "\n",
      "Node:  Brain Volume Blueprint Building Workflow\n",
      "\t [R] Brain Volume Blueprint Building Workflow is a Blueprint Building Workflow.\n",
      "\t [R] Brain Volume Blueprint Building Workflow generates Brain Volume Blueprint.\n",
      "\t [R] Brain Volume Blueprint Building Workflow uses Cell Density.\n",
      "\t [R] Brain Volume Blueprint Building Workflow uses Cell Type Mapping.\n",
      "\n",
      "Node:  Blueprint Building Workflow\n",
      "\t [R] Blueprint Building Workflow is a Workflow.\n",
      "\t [R] Blueprint Building Workflow generates Brain Component Blueprint.\n",
      "\n",
      "Node:  Brain Volume Blueprint\n",
      "\t [R] Brain Volume Blueprint is a Brain Component Blueprint.\n",
      "\n",
      "Node:  Cell Density\n",
      "\t [R] Cell Density is a Brain Volume Feature.\n",
      "\t [R] Cell Density is a Model Brain Volume Parameter.\n",
      "\t [R] Cell Density brainLocation Brain Location.\n",
      "\t [R] Cell Density cellType Brain Cell Type.\n",
      "\n",
      "Node:  Cell Type Mapping\n",
      "\t [R] Cell Type Mapping is a Entity.\n",
      "\t [R] Cell Type Mapping sourceType Brain Cell Type.\n",
      "\t [R] Cell Type Mapping targetType Brain Cell Type.\n",
      "\n",
      "Node:  Subcellular Part Feature\n",
      "\t [D] Subcellular Part Feature is Feature of a subcellular component of a biological brain cell.\n",
      "\t [R] Subcellular Part Feature is a Brain Component Feature.\n",
      "\n",
      "Node:  Model Brain Component\n",
      "\t [D] Model Brain Component is A Model Brain Component is a digital replica of a Biological Brain Component.\n",
      "\t [S] Model Brain Component is also know as Model Brain Component.\n",
      "\t [S] Model Brain Component is referred to as Model Brain Component.\n",
      "\t [R] Model Brain Component is a Entity.\n",
      "\t [R] Model Brain Component is a model of Biological Brain Component.\n",
      "\t [R] Model Brain Component is a model of Brain Component Type.\n",
      "\n",
      "Node:  Ion Channel\n",
      "\t [D] Ion Channel is Ion channels are proteins in the membrane that provide a passageway for charged ions..\n",
      "\t [R] Ion Channel is a Biological Brain Component.\n",
      "\t [R] Ion Channel is a part of Subcellular Part.\n",
      "\t [R] Ion Channel associatedIon Ion.\n",
      "\n",
      "Node:  Workflow\n",
      "\t [D] Workflow is A workflow represents a set of actions or steps intended by one or more agents to achieve some goals. A workflow is a specification or a plan and is instantiated by activities..\n",
      "\t [R] Workflow is a Entity.\n",
      "\t [R] Workflow uses Entity.\n",
      "\t [R] Workflow generates Entity.\n",
      "\t [R] Workflow about Entity.\n",
      "\t [R] Workflow is a part of Workflow.\n",
      "\t [R] Workflow hasWorkflowDefinition WorkflowDefinition.\n",
      "\t [R] Workflow isAssociatedWith SoftwareAgent.\n",
      "\n",
      "Node:  Brain Component Blueprint\n",
      "\t [R] Brain Component Blueprint is a Entity.\n",
      "\t [R] Brain Component Blueprint is a model of Biological Brain Component.\n",
      "\n",
      "Node:  Synapse Plasticity Type\n",
      "\t [R] Synapse Plasticity Type is a Synapse Type.\n",
      "\n",
      "Node:  Synapse Type\n",
      "\t [R] Synapse Type is a Brain Component Type.\n",
      "\t [R] Synapse Type has feature Synapse Feature.\n",
      "\n",
      "Node:  Astrocyte\n",
      "\t [S] Astrocyte is also know as Astrocyte.\n",
      "\t [S] Astrocyte is referred to as Astrocyte.\n",
      "\t [R] Astrocyte is a Glia.\n",
      "\n",
      "Node:  Glia\n",
      "\t [S] Glia is also know as Glia.\n",
      "\t [S] Glia is referred to as Glia.\n",
      "\t [R] Glia is a Brain Cell.\n",
      "\t [R] Glia is a part of NGV Unit.\n",
      "\n",
      "Node:  GABA\n",
      "\t [R] GABA is a Classical Neurotransmitter.\n",
      "\t [R] GABA is a Inhibitory Transmitter.\n",
      "\n",
      "Node:  Classical Neurotransmitter\n",
      "\t [R] Classical Neurotransmitter is a Neurotransmitter.\n",
      "\n",
      "Node:  Inhibitory Transmitter\n",
      "\t [R] Inhibitory Transmitter is a Neurotransmitter.\n",
      "\n",
      "Node:  SoftwareAgent\n",
      "\t [D] SoftwareAgent is ['A software agent is running software.', 'A software application'].\n",
      "\t [R] SoftwareAgent is a Brain Modeling Agent.\n",
      "\t [R] SoftwareAgent algorithm Algorithm.\n",
      "\n",
      "Node:  Brain Modeling Agent\n",
      "\t [D] Brain Modeling Agent is A Brain Modeling Agent is an organization, a person or a software agent that is engaged in the process of digitally reconstructing brain tissue..\n",
      "\n",
      "Node:  Algorithm\n",
      "\t [D] Algorithm is In mathematics and computer science, an algorithm is a finite sequence of well-defined, computer-implementable instructions, typically to solve a class of specific problems or to perform a computation..\n",
      "\t [R] Algorithm is a Entity.\n",
      "\t [R] Algorithm has parameter Model Brain Parameter.\n",
      "\n",
      "Node:  Number Of Synapses Per Connection\n",
      "\t [R] Number Of Synapses Per Connection is a Model Connectivity Parameter.\n",
      "\t [R] Number Of Synapses Per Connection is a Synaptic Pathway Feature.\n",
      "\n",
      "Node:  Model Connectivity Parameter\n",
      "\t [R] Model Connectivity Parameter is a Model Brain Parameter.\n",
      "\n",
      "Node:  Synaptic Pathway Feature\n",
      "\t [R] Synaptic Pathway Feature is a Brain Component Feature.\n",
      "\n",
      "Node:  Neuron Feature\n",
      "\t [D] Neuron Feature is A measurable property or feature of a biological or model neuron..\n",
      "\t [R] Neuron Feature is a Brain Cell Feature.\n",
      "\n",
      "Node:  Brain Cell Feature\n",
      "\t [D] Brain Cell Feature is A measurable property or feature of a biological or model brain cell..\n",
      "\t [R] Brain Cell Feature is a Brain Component Feature.\n",
      "\n",
      "Node:  Synaptic Pathway\n",
      "\t [D] Synaptic Pathway is A set of synaptic connections from one cell type to another cell type..\n",
      "\t [R] Synaptic Pathway is a Biological Brain Component.\n",
      "\t [R] Synaptic Pathway has type Synaptic Pathway Type.\n",
      "\t [R] Synaptic Pathway has feature Synaptic Pathway Feature.\n",
      "\n",
      "Node:  Biological Brain Component\n",
      "\t [D] Biological Brain Component is A Biological Brain Component is a biological entity part of a brain.\n",
      "\t [E] Examples of Biological Brain Component include A Brain Region is a brain component, A Neuron is a brain component, Synaptic Conductance is not a brain component.\n",
      "\t [E] A Brain Region is a brain component, A Neuron is a brain component, Synaptic Conductance is not a brain component is an example of Biological Brain Component.\n",
      "\t [E] Different examples of Biological Brain Component such as A Brain Region is a brain component, A Neuron is a brain component, Synaptic Conductance is not a brain component.\n",
      "\t [R] Biological Brain Component is a Entity.\n",
      "\t [R] Biological Brain Component componentType Brain Component Type.\n",
      "\t [R] Biological Brain Component has feature Brain Component Feature.\n",
      "\n",
      "Node:  Synaptic Pathway Type\n",
      "\t [R] Synaptic Pathway Type is a Brain Component Type.\n",
      "\t [R] Synaptic Pathway Type postsynapticNeuronType Neuron Type.\n",
      "\t [R] Synaptic Pathway Type presynapticNeuronType Neuron Type.\n",
      "\t [R] Synaptic Pathway Type has feature Synaptic Pathway Feature.\n",
      "\n",
      "Node:  Acetylcholine\n",
      "\t [D] Acetylcholine is An ester of choline and acetic acid that serves as a transmitter substance of nerve impulses within the central and peripheral nervous systems. Acetylcholine is the chief neurotransmitter of the parasympathetic nervous system, the part of the autonomic nervous system (a branch of the peripheral nervous system) that contracts smooth muscles, dilates blood vessels, increases bodily secretions, and slows heart rate. Acetylcholine can stimulate a response or block a response and thus can have excitatory or inhibitory effects..\n",
      "\t [R] Acetylcholine is a Neuromodulator.\n",
      "\n",
      "Node:  Neuromodulator\n",
      "\t [D] Neuromodulator is A chemical uses by neurons to regulate diverse populations of neurons. Neuromodulators typically bind to metabotropic, G-protein coupled receptors (GPCRs) to initiate a second messenger signaling cascade that induces a broad, long-lasting signal. This modulation can last for hundreds of milliseconds to several minutes. Some of the effects of neuromodulators include: alter intrinsic firing activity, increase or decrease voltage-dependent currents, alter synaptic efficacy, increase bursting activity and reconfiguration of synaptic connectivity. Major neuromodulators in the central nervous system include: dopamine, serotonin, acetylcholine, histamine, norepinephrine, nitric oxide, and several neuropeptides..\n",
      "\t [R] Neuromodulator is a Neurotransmitter.\n",
      "\n",
      "Node:  Model Reaction Validation Workflow\n",
      "\t [R] Model Reaction Validation Workflow is a Model Validation Workflow.\n",
      "\t [R] Model Reaction Validation Workflow hasTarget Steady State Molecular Concentration.\n",
      "\t [R] Model Reaction Validation Workflow uses Molecular Concentration Series.\n",
      "\n",
      "Node:  Model Validation Workflow\n",
      "\t [R] Model Validation Workflow is a Workflow.\n",
      "\t [R] Model Validation Workflow hasTarget Brain Component Feature.\n",
      "\t [R] Model Validation Workflow uses Dataset.\n",
      "\n",
      "Node:  Steady State Molecular Concentration\n",
      "\t [R] Steady State Molecular Concentration is a Molecular Compartment Feature.\n",
      "\t [R] Steady State Molecular Concentration molecule Molecule.\n",
      "\n",
      "Node:  Molecular Concentration Series\n",
      "\t [D] Molecular Concentration Series is Time series representing change of molecular concentration value of a given molecule in a given molecular compartment..\n",
      "\t [R] Molecular Concentration Series is a Time Series.\n",
      "\n",
      "Node:  Histamine\n",
      "\t [D] Histamine is Histamine is an organic nitrogenous compound involved in local immune responses, as well as regulating physiological functions in the gut and acting as a neurotransmitter for the brain, spinal cord, and uterus. In recent years, histamine has been recognized as a central neurotransmitter. It consists of an imidazole ring attached to an ethylamine chain; under physiological conditions, the amino group of the side-chain is protonated..\n",
      "\t [R] Histamine is a Excitatory Transmitter.\n",
      "\t [R] Histamine is a Neuromodulator.\n",
      "\n",
      "Node:  Excitatory Transmitter\n",
      "\t [R] Excitatory Transmitter is a Neurotransmitter.\n",
      "\n",
      "Node:  Brain Area\n",
      "\t [D] Brain Area is A Brain Area is a part of a Brain Region that is anatomically and functionally distinguishable. A Brain Region is made up of multiple Brain Areas..\n",
      "\t [E] Examples of Brain Area include An example of Brain Areas include V1, V2, V3,V4, etc of the Visual Cortex Brain Region or S1,S2, etc of the Somatosensory Cortex Brain Region.\n",
      "\t [E] An example of Brain Areas include V1, V2, V3,V4, etc of the Visual Cortex Brain Region or S1,S2, etc of the Somatosensory Cortex Brain Region is an example of Brain Area.\n",
      "\t [E] Different examples of Brain Area such as An example of Brain Areas include V1, V2, V3,V4, etc of the Visual Cortex Brain Region or S1,S2, etc of the Somatosensory Cortex Brain Region.\n",
      "\t [R] Brain Area is a Biological Brain Component.\n",
      "\t [R] Brain Area is a Brain Volume.\n",
      "\t [R] Brain Area is a part of Brain Region.\n",
      "\n",
      "Node:  Brain Volume\n",
      "\t [D] Brain Volume is Any bounded space in the brain where neurons can reside.\n",
      "\t [R] Brain Volume is a Biological Brain Component.\n",
      "\n",
      "Node:  Brain Region\n",
      "\t [D] Brain Region is Anatomical divisons of the brain according to one or more criteria, e.g. cytoarchitectural, gross anatomy. Parts may be contiguous in 3D or not, e.g., basal ganglia..\n",
      "\t [S] Brain Region is also know as regional part of the brain.\n",
      "\t [S] Brain Region is referred to as regional part of the brain.\n",
      "\t [R] Brain Region is a Biological Brain Component.\n",
      "\t [R] Brain Region is a Brain Volume.\n",
      "\t [R] Brain Region is a part of Brain System.\n",
      "\t [R] Brain Region species Species.\n",
      "\n",
      "Node:  Model Biochemical Reaction Parameter\n",
      "\t [D] Model Biochemical Reaction Parameter is Parameter used to represent a specific biological or theoretical value that is used in the modeling of biochemical reactions..\n",
      "\t [R] Model Biochemical Reaction Parameter is a Model Brain Parameter.\n",
      "\n",
      "Node:  Model Brain Parameter\n",
      "\t [D] Model Brain Parameter is A brain parameter is a parameter used to represent a specific biological or theoretical value that is used in the construction of a brain model..\n",
      "\t [R] Model Brain Parameter is a Entity.\n",
      "\t [R] Model Brain Parameter value Quantitative Value.\n",
      "\n",
      "Node:  Brain Volume Feature\n",
      "\t [R] Brain Volume Feature is a Brain Component Feature.\n",
      "\n",
      "Node:  Modulatory Neuron\n",
      "\t [D] Modulatory Neuron is A nueron that releases modulatory neurotransmitters. These neurotransmitters, often referred to as neuromodulators, are capable of affecting a larger number of neurons at the same time. These neuromodulators also influence the effects of other chemical messengers. Where synaptic neurotransmitters are released by axon terminals to have a fast-acting impact on other receptor neurons, neuromodulators diffuse across a larger area and are more slow-acting..\n",
      "\t [S] Modulatory Neuron is also know as Modulatory Neuron.\n",
      "\t [S] Modulatory Neuron is referred to as Modulatory Neuron.\n",
      "\t [R] Modulatory Neuron is a Neurotransmitter Type.\n",
      "\n",
      "Node:  Neurotransmitter Type\n",
      "\t [S] Neurotransmitter Type is also know as Neurotransmitter Type.\n",
      "\t [S] Neurotransmitter Type is referred to as Neurotransmitter Type.\n",
      "\t [R] Neurotransmitter Type is a Neuron Type.\n",
      "\n",
      "Node:  Inhibitory Neuron\n",
      "\t [D] Inhibitory Neuron is A neuron that releases inhibitory neurotransmitters. These types of neurotransmitters have inhibitory effects on the neuron; they decrease the likelihood that the neuron will fire an action potential. Some of the major inhibitory neurotransmitters include serotonin and gamma-aminobutyric acid (GABA)..\n",
      "\t [S] Inhibitory Neuron is also know as Inhibitory Neuron.\n",
      "\t [S] Inhibitory Neuron is referred to as Inhibitory Neuron.\n",
      "\t [R] Inhibitory Neuron is a Neurotransmitter Type.\n",
      "\n",
      "Node:  Glia Metabolism Building Workflow\n",
      "\t [R] Glia Metabolism Building Workflow is a Glia Building Workflow.\n",
      "\t [R] Glia Metabolism Building Workflow generates Model Glia Metabolism.\n",
      "\t [R] Glia Metabolism Building Workflow uses Model Biochemical Reaction.\n",
      "\t [R] Glia Metabolism Building Workflow uses Model Subcellular Part Metabolism.\n",
      "\n",
      "Node:  Glia Building Workflow\n",
      "\t [R] Glia Building Workflow is a Cell Building Workflow.\n",
      "\t [R] Glia Building Workflow generates Model Glia.\n",
      "\t [R] Glia Building Workflow about Glia.\n",
      "\n",
      "Node:  Model Glia Metabolism\n",
      "\t [R] Model Glia Metabolism is a Model Glia.\n",
      "\t [R] Model Glia Metabolism is a part of Model NGV Unit Metabolism.\n",
      "\t [R] Model Glia Metabolism is a model of Glia.\n",
      "\n",
      "Node:  Model Biochemical Reaction\n",
      "\t [D] Model Biochemical Reaction is A computational model of a biochemical reaction, a chemical process of transformation of some molecules (substrates) into the other molecules (products)..\n",
      "\t [R] Model Biochemical Reaction is a Model Brain Activity.\n",
      "\t [R] Model Biochemical Reaction is a model of Biochemical Reaction.\n",
      "\n",
      "Node:  Model Neuron\n",
      "\t [R] Model Neuron is a Model Cell.\n",
      "\t [R] Model Neuron is a model of Neuron.\n",
      "\n",
      "Node:  Model Cell\n",
      "\t [R] Model Cell is a Model Brain Component.\n",
      "\t [R] Model Cell is a model of Brain Cell.\n",
      "\n",
      "Node:  Neuron\n",
      "\t [R] Neuron is a Brain Cell.\n",
      "\t [R] Neuron is a part of NGV Unit.\n",
      "\t [R] Neuron has type Neuron Type.\n",
      "\t [R] Neuron has feature Neuron Feature.\n",
      "\n",
      "Node:  IBEA\n",
      "\t [D] IBEA is Indicator-based evolutionary algorithm..\n",
      "\t [R] IBEA is a EModel Parameter Optimizer.\n",
      "\t [R] IBEA is a Evolutionary Algorithm.\n",
      "\n",
      "Node:  EModel Parameter Optimizer\n",
      "\t [D] EModel Parameter Optimizer is Algorithm used for optimization..\n",
      "\t [R] EModel Parameter Optimizer is a EModel Hyperparameter.\n",
      "\n",
      "Node:  Evolutionary Algorithm\n",
      "\t [D] Evolutionary Algorithm is In computational intelligence, an evolutionary algorithm is a subset of evolutionary computation, a generic population-based metaheuristic optimization algorithm..\n",
      "\t [R] Evolutionary Algorithm is a Algorithm.\n",
      "\n",
      "Node:  K\n",
      "\t [S] K is also know as Potassium.\n",
      "\t [S] K is referred to as Potassium.\n",
      "\t [R] K is a Ion.\n",
      "\n",
      "Node:  Ion\n",
      "\t [D] Ion is An ion is a particle, atom or molecule with a net electrical charge. The movement of ions through the membrane of neurons underlies their signalling mechanism..\n",
      "\t [R] Ion is a Biological Brain Component.\n",
      "\n",
      "Node:  Synaptic Connection\n",
      "\t [D] Synaptic Connection is A set of synapses connecting one concrete cell to another..\n",
      "\t [R] Synaptic Connection is a Biological Brain Component.\n",
      "\t [R] Synaptic Connection is a part of Synaptic Pathway.\n",
      "\t [R] Synaptic Connection postsynapticNeuron Neuron.\n",
      "\t [R] Synaptic Connection presynapticNeuron Neuron.\n",
      "\t [R] Synaptic Connection has feature Synaptic Connection Feature.\n",
      "\n",
      "Node:  Vector3D\n",
      "\t [R] Vector3D is a Entity.\n",
      "\n",
      "Node:  Time Series\n",
      "\t [R] Time Series is a Dataset.\n",
      "\n",
      "Node:  Dataset\n",
      "\t [R] Dataset is a Entity.\n",
      "\t [R] Dataset brainLocation Brain Location.\n",
      "\t [R] Dataset generationModality Data Generation Modality.\n",
      "\t [R] Dataset object Object Of Study.\n",
      "\t [R] Dataset subject Subject.\n",
      "\n",
      "Node:  Synapse\n",
      "\t [D] Synapse is A structure that permits a neuron to pass an electrical or chemical signal to another neuron or to the target effector cell. Synapse is formed by a presynaptic and postsynaptic part with a space between the two parts. Different structural proteins connect the pre- and postsynaptic parts of the synapse..\n",
      "\t [R] Synapse is a Biological Brain Component.\n",
      "\t [R] Synapse is a part of Synaptic Connection.\n",
      "\t [R] Synapse has type Synapse Type.\n",
      "\t [R] Synapse postsynapticNeuron Neuron.\n",
      "\t [R] Synapse presynapticNeuron Neuron.\n",
      "\t [R] Synapse has feature Synapse Feature.\n",
      "\n",
      "Node:  Synapse Feature\n",
      "\t [R] Synapse Feature is a Brain Component Feature.\n",
      "\n",
      "Node:  Model Brain Vascular Component\n",
      "\t [R] Model Brain Vascular Component is a Model Brain Component.\n",
      "\t [R] Model Brain Vascular Component is a model of Brain Vascular Component.\n",
      "\n",
      "Node:  Brain Vascular Component\n",
      "\t [D] Brain Vascular Component is A Brain Vascular Component is a biological entity part of the brain vasculature..\n",
      "\t [E] Examples of Brain Vascular Component include A brain artery is an example of a brain vascular component..\n",
      "\t [E] A brain artery is an example of a brain vascular component. is an example of Brain Vascular Component.\n",
      "\t [E] Different examples of Brain Vascular Component such as A brain artery is an example of a brain vascular component..\n",
      "\t [R] Brain Vascular Component is a Biological Brain Component.\n",
      "\t [R] Brain Vascular Component is a Molecular Compartment.\n",
      "\t [R] Brain Vascular Component is a part of Microcircuit.\n",
      "\t [R] Brain Vascular Component is a part of NGV Unit.\n",
      "\n",
      "Node:  Basal Dendrite\n",
      "\t [R] Basal Dendrite is a Dendrite.\n",
      "\n",
      "Node:  Dendrite\n",
      "\t [R] Dendrite is a Neuron Part.\n",
      "\t [R] Dendrite is a part of Neuron.\n",
      "\n",
      "Node:  Neurotransmitter Synapse Type\n",
      "\t [R] Neurotransmitter Synapse Type is a Synapse Type.\n",
      "\n",
      "Node:  Cell Building Workflow\n",
      "\t [R] Cell Building Workflow is a Model Building Workflow.\n",
      "\t [R] Cell Building Workflow generates Model Cell.\n",
      "\t [R] Cell Building Workflow about Brain Cell.\n",
      "\n",
      "Node:  Model Glia\n",
      "\t [R] Model Glia is a Model Cell.\n",
      "\t [R] Model Glia is a model of Glia.\n",
      "\n",
      "Node:  Homo Sapiens\n",
      "\t [S] Homo Sapiens is also know as Human.\n",
      "\t [S] Homo Sapiens is referred to as Human.\n",
      "\t [R] Homo Sapiens is a Species.\n",
      "\n",
      "Node:  Species\n",
      "\t [D] Species is In biology, a species is a basic unit of classification and a taxonomic rank of an organism, as well as a unit of biodiversity..\n",
      "\t [R] Species is a Entity.\n",
      "\n",
      "Node:  Activity\n",
      "\t [D] Activity is An activity is something that occurs over a period of time and acts upon or with entities; it may include consuming, processing, transforming, modifying, relocating, using, or generating entities..\n",
      "\t [R] Activity generated Entity.\n",
      "\t [R] Activity used Entity.\n",
      "\t [R] Activity hadPlan Workflow.\n",
      "\t [R] Activity wasAssociatedWith Brain Modeling Agent.\n",
      "\n",
      "Node:  Brain Component Type\n",
      "\t [D] Brain Component Type is Classification of Brain Components whether biological or model..\n",
      "\t [R] Brain Component Type is a Entity.\n",
      "\t [R] Brain Component Type has feature Brain Component Feature.\n",
      "\n",
      "Node:  Gamma CaDynamics DC0\n",
      "\t [D] Gamma CaDynamics DC0 is The inverse of the calcium ion buffer's binding ratio..\n",
      "\t [R] Gamma CaDynamics DC0 is a Calcium Dynamics.\n",
      "\n",
      "Node:  Calcium Dynamics\n",
      "\t [D] Calcium Dynamics is Kinetics of the conductance mechanism of calcium ion..\n",
      "\t [R] Calcium Dynamics is a Ion Dynamics Feature.\n",
      "\t [R] Calcium Dynamics ion Ca.\n",
      "\n",
      "Node:  Model Basal Lamina Metabolism\n",
      "\t [R] Model Basal Lamina Metabolism is a Model Extracellular Part Metabolism.\n",
      "\t [R] Model Basal Lamina Metabolism is a model of Basal Lamina.\n",
      "\n",
      "Node:  Model Extracellular Part Metabolism\n",
      "\t [R] Model Extracellular Part Metabolism is a Model Extracellular Part.\n",
      "\t [R] Model Extracellular Part Metabolism is a part of Model NGV Unit Metabolism.\n",
      "\t [R] Model Extracellular Part Metabolism is a model of Extracellular Part.\n",
      "\n",
      "Node:  Basal Lamina\n",
      "\t [D] Basal Lamina is A Basal Lamina is a specialised sheet of extracellular matrix..\n",
      "\t [R] Basal Lamina is a Extracellular Part.\n",
      "\n",
      "Node:  Neuron Type\n",
      "\t [R] Neuron Type is a Brain Cell Type.\n",
      "\t [R] Neuron Type has feature Neuron Feature.\n",
      "\n",
      "Node:  Neurotransmitter\n",
      "\t [D] Neurotransmitter is Blue Brain considers a neurotransmitter as any chemical agent that is produced by one cell in any part of the body and targets one or more cells in the Brain.\n",
      "\t [R] Neurotransmitter is a Biological Brain Component.\n",
      "\n",
      "Node:  BACFiringFeature\n",
      "\t [D] BACFiringFeature is Back-propagating action potential activated Ca2+ spike firing feature..\n",
      "\t [R] BACFiringFeature is a Neuron Electrophysiological Feature.\n",
      "\t [R] BACFiringFeature compartment Soma.\n",
      "\t [R] BACFiringFeature compartment Dendrite.\n",
      "\n",
      "Node:  Soma\n",
      "\t [R] Soma is a Neuron Part.\n",
      "\t [R] Soma is a part of Neuron.\n",
      "\n",
      "Node:  Neuron Projection Type\n",
      "\t [S] Neuron Projection Type is also know as PType.\n",
      "\t [S] Neuron Projection Type is referred to as PType.\n",
      "\t [R] Neuron Projection Type is a Neuron Type.\n",
      "\n",
      "Node:  Leaf Regions\n",
      "\t [R] Leaf Regions is a Neurite Location Feature.\n",
      "\n",
      "Node:  Neurite Location Feature\n",
      "\t [R] Neurite Location Feature is a Neuron Feature.\n",
      "\n",
      "Node:  Brain System\n",
      "\t [D] Brain System is A Brain system is a system made up of at least two brain regions, including the anatomical and/or functional connections  between these brain regions..\n",
      "\t [E] Examples of Brain System include Neocortical Thalamic System.\n",
      "\t [E] Neocortical Thalamic System is an example of Brain System.\n",
      "\t [E] Different examples of Brain System such as Neocortical Thalamic System.\n",
      "\t [R] Brain System is a Biological Brain Component.\n",
      "\t [R] Brain System is a Brain Volume.\n",
      "\t [R] Brain System is a part of Whole Brain.\n",
      "\n",
      "Node:  Whole Brain\n",
      "\t [D] Whole Brain is The brain is the center of the nervous system in all vertebrate, and most invertebrate, animals. Some primitive animals such as jellyfish and starfish have a decentralized nervous system without a brain, while sponges lack any nervous system at all. In vertebrates, the brain is located in the head, protected by the skull and close to the primary sensory apparatus of vision, hearing, balance, taste, and smell.\n",
      "\t [R] Whole Brain is a Biological Brain Component.\n",
      "\t [R] Whole Brain is a Brain Volume.\n",
      "\n",
      "Node:  Angiotensin\n",
      "\t [R] Angiotensin is a Hormone.\n",
      "\n",
      "Node:  Hormone\n",
      "\t [R] Hormone is a Neurotransmitter.\n",
      "\n",
      "Node:  Brain Cell\n",
      "\t [D] Brain Cell is Blue Brain considers as brain cells neurons, glia, endothelial cells as well as externally invading cells e.g. from the immune system.\n",
      "\t [R] Brain Cell is a Biological Brain Component.\n",
      "\t [R] Brain Cell is a Molecular Compartment.\n",
      "\t [R] Brain Cell is a part of Microcircuit.\n",
      "\t [R] Brain Cell has type Brain Cell Type.\n",
      "\t [R] Brain Cell has feature Brain Cell Feature.\n",
      "\n",
      "Node:  Molecular Compartment\n",
      "\t [D] Molecular Compartment is A compartment in which molecular species live and in which biochemical reactions can take place..\n",
      "\t [R] Molecular Compartment is a Biological Brain Component.\n",
      "\t [R] Molecular Compartment has feature Molecular Compartment Feature.\n",
      "\n",
      "Node:  Microcircuit\n",
      "\t [D] Microcircuit is A microcircuit is a minimal volume of brain tissue where the cell numbers are sufficient to saturate the volume of the dendritic neuropil. A number of microcircuits make up a Brain Area..\n",
      "\t [S] Microcircuit is also know as Neuronal Microcircuit.\n",
      "\t [S] Microcircuit is referred to as Neuronal Microcircuit.\n",
      "\t [S] Microcircuit is also know as Microcircuit.\n",
      "\t [S] Microcircuit is referred to as Microcircuit.\n",
      "\t [R] Microcircuit is a Biological Brain Component.\n",
      "\t [R] Microcircuit is a Brain Volume.\n",
      "\t [R] Microcircuit is a part of Brain Area.\n",
      "\n",
      "Node:  Brain Cell Type\n",
      "\t [R] Brain Cell Type is a Brain Component Type.\n",
      "\t [R] Brain Cell Type has feature Brain Cell Feature.\n",
      "\n",
      "Node:  Brain Modeling Activity\n",
      "\t [D] Brain Modeling Activity is A Brain Modeling Activity is any activity involved the process of digitally reconstructing brain tissue. This includes but not limited to collecting values from literature, db and generated from experiment..\n",
      "\t [R] Brain Modeling Activity is a Activity.\n",
      "\n",
      "Node:  Model Circuit Building Workflow\n",
      "\t [R] Model Circuit Building Workflow is a Model Building Workflow.\n",
      "\t [R] Model Circuit Building Workflow generates Model Circuit.\n",
      "\t [R] Model Circuit Building Workflow uses Model Brain Volume.\n",
      "\t [R] Model Circuit Building Workflow uses Model Circuit Connectivity.\n",
      "\n",
      "Node:  Model Building Workflow\n",
      "\t [R] Model Building Workflow is a Workflow.\n",
      "\t [R] Model Building Workflow uses Model Brain Component.\n",
      "\t [R] Model Building Workflow generates Model Brain Component.\n",
      "\t [R] Model Building Workflow has parameter Model Brain Parameter.\n",
      "\t [R] Model Building Workflow isAssociatedWith SoftwareAgent.\n",
      "\t [R] Model Building Workflow hasTarget Brain Component Feature.\n",
      "\t [R] Model Building Workflow uses Dataset.\n",
      "\t [R] Model Building Workflow uses Equation.\n",
      "\n",
      "Node:  Model Circuit\n",
      "\t [R] Model Circuit is a Model Brain Component.\n",
      "\t [R] Model Circuit is a model of Brain Volume.\n",
      "\n",
      "Node:  Model Brain Volume\n",
      "\t [R] Model Brain Volume is a Model Brain Component.\n",
      "\t [R] Model Brain Volume is a model of Brain Volume.\n",
      "\n",
      "Node:  Model Circuit Connectivity\n",
      "\t [R] Model Circuit Connectivity is a Model Brain Component.\n",
      "\t [R] Model Circuit Connectivity is a model of Brain Volume.\n",
      "\n",
      "Node:  Model Interstitial Space Metabolism\n",
      "\t [R] Model Interstitial Space Metabolism is a Model Extracellular Part Metabolism.\n",
      "\t [R] Model Interstitial Space Metabolism is a model of Interstitial Space.\n",
      "\n",
      "Node:  Interstitial Space\n",
      "\t [D] Interstitial Space is The Interstitial Space is a narrow space between neural cells and blood vessels..\n",
      "\t [R] Interstitial Space is a Extracellular Part.\n",
      "\n",
      "Node:  Extracellular Part\n",
      "\t [D] Extracellular Part is An Extracellular Part is a biological brain component situated outside brain cells..\n",
      "\t [E] Examples of Extracellular Part include The brain's extracellular matrix is an extracellular part..\n",
      "\t [E] The brain's extracellular matrix is an extracellular part. is an example of Extracellular Part.\n",
      "\t [E] Different examples of Extracellular Part such as The brain's extracellular matrix is an extracellular part..\n",
      "\t [R] Extracellular Part is a Biological Brain Component.\n",
      "\t [R] Extracellular Part is a Molecular Compartment.\n",
      "\n",
      "Node:  NGV Unit Metabolism Building Workflow\n",
      "\t [R] NGV Unit Metabolism Building Workflow is a NGV Unit Building Workflow.\n",
      "\t [R] NGV Unit Metabolism Building Workflow generates Model NGV Unit Metabolism.\n",
      "\t [R] NGV Unit Metabolism Building Workflow about NGV Unit.\n",
      "\t [R] NGV Unit Metabolism Building Workflow uses Model Basal Lamina Metabolism.\n",
      "\t [R] NGV Unit Metabolism Building Workflow uses Model Biochemical Reaction.\n",
      "\t [R] NGV Unit Metabolism Building Workflow uses Model Glia Metabolism.\n",
      "\t [R] NGV Unit Metabolism Building Workflow uses Model Interstitial Space Metabolism.\n",
      "\t [R] NGV Unit Metabolism Building Workflow uses Model Neuron Metabolism.\n",
      "\n",
      "Node:  NGV Unit Building Workflow\n",
      "\t [R] NGV Unit Building Workflow is a Model Building Workflow.\n",
      "\t [R] NGV Unit Building Workflow generates Model NGV Unit.\n",
      "\t [R] NGV Unit Building Workflow about NGV Unit.\n",
      "\n",
      "Node:  Model NGV Unit Metabolism\n",
      "\t [D] Model NGV Unit Metabolism is Model of a metabolic system of a Neuron/Glia/Vasculature unit.\n",
      "\t [R] Model NGV Unit Metabolism is a Model NGV Unit.\n",
      "\t [R] Model NGV Unit Metabolism is a model of NGV Unit.\n",
      "\n",
      "Node:  NGV Unit\n",
      "\t [D] NGV Unit is An NGV Unit at Blue Brain refers to a \"minimal\" neuro-glia-vasculatur unit including metabolism of a capillary, neuron and astrocyte. The Molecular Systems team at Blue Brain reconstructs and simulates such an NGV Unit. Multiple NGV Units are put together to form an NGV Column which is used to enhance the neocortical column with metabolism..\n",
      "\t [R] NGV Unit is a Biological Brain Component.\n",
      "\t [R] NGV Unit is a Molecular Compartment.\n",
      "\n",
      "Node:  Model Neuron Metabolism\n",
      "\t [D] Model Neuron Metabolism is Model of a metabolic network of a neuron..\n",
      "\t [R] Model Neuron Metabolism is a Model Neuron.\n",
      "\t [R] Model Neuron Metabolism is a part of Model NGV Unit Metabolism.\n",
      "\t [R] Model Neuron Metabolism is a model of Neuron.\n",
      "\n",
      "Node:  I2 Synapse Type\n",
      "\t [R] I2 Synapse Type is a Depressing Synapse Type.\n",
      "\t [R] I2 Synapse Type is a Inhibitory Synapse Type.\n",
      "\n",
      "Node:  Depressing Synapse Type\n",
      "\t [R] Depressing Synapse Type is a Synapse Plasticity Type.\n",
      "\n",
      "Node:  Inhibitory Synapse Type\n",
      "\t [R] Inhibitory Synapse Type is a Neurotransmitter Synapse Type.\n",
      "\n",
      "Node:  EModel Parameter Constraint Generation Workflow\n",
      "\t [R] EModel Parameter Constraint Generation Workflow is a Parameter Constraint Generation Workflow.\n",
      "\t [R] EModel Parameter Constraint Generation Workflow generates EModel Parameter Constraint.\n",
      "\n",
      "Node:  Parameter Constraint Generation Workflow\n",
      "\t [R] Parameter Constraint Generation Workflow is a Workflow.\n",
      "\t [R] Parameter Constraint Generation Workflow generates Model Brain Parameter Constraint.\n",
      "\n",
      "Node:  EModel Parameter Constraint\n",
      "\t [R] EModel Parameter Constraint is a Model Brain Parameter Constraint.\n",
      "\n",
      "Node:  EModel Simulation Workflow\n",
      "\t [R] EModel Simulation Workflow is a Brain Component Simulation Workflow.\n",
      "\t [R] EModel Simulation Workflow generates Trace.\n",
      "\t [R] EModel Simulation Workflow uses EModel.\n",
      "\n",
      "Node:  Brain Component Simulation Workflow\n",
      "\t [R] Brain Component Simulation Workflow is a Workflow.\n",
      "\t [R] Brain Component Simulation Workflow generates Dataset.\n",
      "\t [R] Brain Component Simulation Workflow about Biological Brain Component.\n",
      "\t [R] Brain Component Simulation Workflow uses Model Brain Component.\n",
      "\n",
      "Node:  Trace\n",
      "\t [R] Trace is a Electrical Series.\n",
      "\t [R] Trace hasMType Neuron Morphological Type.\n",
      "\n",
      "Node:  EModel\n",
      "\t [R] EModel is a Model Neuron.\n",
      "\t [R] EModel is a model of Neuron Electrical Type.\n",
      "\n",
      "Node:  EModel Parameter\n",
      "\t [R] EModel Parameter is a Model Brain Parameter.\n",
      "\n",
      "Node:  Neuron Part Feature\n",
      "\t [R] Neuron Part Feature is a Subcellular Part Feature.\n",
      "\n",
      "Node:  Neuron Part\n",
      "\t [R] Neuron Part is a Subcellular Part.\n",
      "\t [R] Neuron Part is a part of Neuron.\n",
      "\t [R] Neuron Part has feature Neuron Part Feature.\n",
      "\n",
      "Node:  Electrical Stimulus\n",
      "\t [R] Electrical Stimulus is a Entity.\n",
      "\n",
      "Node:  Model Extracellular Part\n",
      "\t [R] Model Extracellular Part is a Model Brain Component.\n",
      "\t [R] Model Extracellular Part is a model of Extracellular Part.\n",
      "\n",
      "Node:  Persistent K+ Current\n",
      "\t [D] Persistent K+ Current is The persistent component of the potassium current.\n",
      "\t [S] Persistent K+ Current is also know as Persistent potassium current.\n",
      "\t [S] Persistent K+ Current is referred to as Persistent potassium current.\n",
      "\t [S] Persistent K+ Current is also know as K_Pst.\n",
      "\t [S] Persistent K+ Current is referred to as K_Pst.\n",
      "\t [R] Persistent K+ Current is a Ionic Mechanism.\n",
      "\t [R] Persistent K+ Current ion K.\n",
      "\n",
      "Node:  Subcellular Part Metabolism Building Workflow\n",
      "\t [R] Subcellular Part Metabolism Building Workflow is a Subcellular Part Building Workflow.\n",
      "\t [R] Subcellular Part Metabolism Building Workflow generates Model Subcellular Part Metabolism.\n",
      "\t [R] Subcellular Part Metabolism Building Workflow about Subcellular Part.\n",
      "\t [R] Subcellular Part Metabolism Building Workflow uses Model Biochemical Reaction.\n",
      "\n",
      "Node:  Subcellular Part Building Workflow\n",
      "\t [R] Subcellular Part Building Workflow is a Model Building Workflow.\n",
      "\t [R] Subcellular Part Building Workflow generates Model Subcellular Part.\n",
      "\t [R] Subcellular Part Building Workflow about Subcellular Part.\n",
      "\n",
      "Node:  Subcellular Part\n",
      "\t [D] Subcellular Part is A Subcellular Part is a biological brain component with a level of organisation lower than the cellular..\n",
      "\t [E] Examples of Subcellular Part include A mitochondrion is a subcellular part..\n",
      "\t [E] A mitochondrion is a subcellular part. is an example of Subcellular Part.\n",
      "\t [E] Different examples of Subcellular Part such as A mitochondrion is a subcellular part..\n",
      "\t [R] Subcellular Part is a Biological Brain Component.\n",
      "\t [R] Subcellular Part is a Molecular Compartment.\n",
      "\t [R] Subcellular Part is a part of Brain Cell.\n",
      "\t [R] Subcellular Part has feature Subcellular Part Feature.\n",
      "\n",
      "Node:  Model Reaction Parameter Optimization Workflow\n",
      "\t [R] Model Reaction Parameter Optimization Workflow is a Parameter Optimization Workflow.\n",
      "\t [R] Model Reaction Parameter Optimization Workflow is a part of Model Reaction Building Workflow.\n",
      "\t [R] Model Reaction Parameter Optimization Workflow generates Model Biochemical Reaction Parameter.\n",
      "\t [R] Model Reaction Parameter Optimization Workflow hasTarget Steady State Molecular Concentration.\n",
      "\t [R] Model Reaction Parameter Optimization Workflow uses Reaction Rate Equation.\n",
      "\t [R] Model Reaction Parameter Optimization Workflow uses Molecular Concentration.\n",
      "\n",
      "Node:  Model Reaction Building Workflow\n",
      "\t [R] Model Reaction Building Workflow is a Model Building Workflow.\n",
      "\t [R] Model Reaction Building Workflow generates Model Biochemical Reaction.\n",
      "\t [R] Model Reaction Building Workflow has parameter Model Biochemical Reaction Parameter.\n",
      "\t [R] Model Reaction Building Workflow about Biochemical Reaction.\n",
      "\t [R] Model Reaction Building Workflow hasTarget Steady State Molecular Concentration.\n",
      "\t [R] Model Reaction Building Workflow uses Reaction Rate Equation.\n",
      "\t [R] Model Reaction Building Workflow uses Molecular Concentration.\n",
      "\n",
      "Node:  Reaction Rate Equation\n",
      "\t [D] Reaction Rate Equation is A rate equation or rate law is a mathematical expression used in chemical kinetics to link the rate of a reaction to the concentration of each reactant..\n",
      "\t [R] Reaction Rate Equation is a Equation.\n",
      "\n",
      "Node:  Molecular Concentration\n",
      "\t [D] Molecular Concentration is Concentration value of a given molecule in a given molecular compartment.\n",
      "\t [S] Molecular Concentration is also know as Molecular Concentration Value.\n",
      "\t [S] Molecular Concentration is referred to as Molecular Concentration Value.\n",
      "\t [R] Molecular Concentration is a Dataset.\n",
      "\n",
      "Node:  Predicted Data\n",
      "\t [R] Predicted Data is a Data Generation Modality.\n",
      "\n",
      "Node:  Data Generation Modality\n",
      "\t [R] Data Generation Modality is a Entity.\n",
      "\n",
      "Node:  Feature Extraction Workflow\n",
      "\t [R] Feature Extraction Workflow is a Workflow.\n",
      "\t [R] Feature Extraction Workflow is a part of Model Building Workflow.\n",
      "\t [R] Feature Extraction Workflow generates Brain Component Feature.\n",
      "\t [R] Feature Extraction Workflow uses Dataset.\n",
      "\n",
      "Node:  Protein Protein Interaction\n",
      "\t [R] Protein Protein Interaction is a Biochemical Reaction.\n",
      "\n",
      "Node:  Biochemical Reaction\n",
      "\t [D] Biochemical Reaction is A chemical process of transformation of some molecules (substrates) into the other molecules (products)..\n",
      "\t [R] Biochemical Reaction is a Biological Brain Activity.\n",
      "\n",
      "Node:  Neuron Building Workflow\n",
      "\t [R] Neuron Building Workflow is a Cell Building Workflow.\n",
      "\t [R] Neuron Building Workflow generates Model Neuron.\n",
      "\t [R] Neuron Building Workflow about Neuron.\n",
      "\n",
      "Node:  Neuron Electrical Type\n",
      "\t [R] Neuron Electrical Type is a Neuron Type.\n",
      "\t [R] Neuron Electrical Type has feature Neuron Electrophysiological Feature.\n",
      "\n",
      "Node:  Association\n",
      "\t [D] Association is An activity association is an assignment of responsibility to an agent for an activity, indicating that the agent had a role in the activity. It further allows for a plan to be specified, which is the plan intended by the agent to achieve some goals in the context of this activity..\n",
      "\t [R] Association is a Data Provenance.\n",
      "\n",
      "Node:  SO-CMA\n",
      "\t [D] SO-CMA is Single-objective covariance matrix adaptation evolution strategy..\n",
      "\t [R] SO-CMA is a EModel Parameter Optimizer.\n",
      "\t [R] SO-CMA is a Evolutionary Algorithm.\n",
      "\n",
      "Node:  lateral geniculate nucleus parvocellular layer\n",
      "\t [R] lateral geniculate nucleus parvocellular layer is a Brain Area.\n",
      "\n",
      "Node:  Serotonin\n",
      "\t [D] Serotonin is A monoamine neurotransmitter. Its biological function is complex and multifaceted, modulating mood, cognition, reward, learning, memory, and numerous physiological processes such as vomiting and vasoconstriction. Biochemically, the indoleamine molecule derives from the amino acid tryptophan, via the (rate-limiting) hydroxylation of the 5 position on the ring (forming the intermediate 5-hydroxytryptophan), and then decarboxylation to produce serotonin. Serotonin is primarily found in the enteric nervous system located in the gastrointestinal tract (GI tract). However, it is also produced in the central nervous system (CNS), specifically in the raphe nuclei located in the brainstem, Merkel cells located in the skin, pulmonary neuroendocrine cells and taste receptor cells in the tongue. Additionally, serotonin is stored in blood platelets and is released during agitation and vasoconstriction, where it then acts as an agonist to other platelets..\n",
      "\t [R] Serotonin is a Excitatory Transmitter.\n",
      "\t [R] Serotonin is a Neuromodulator.\n",
      "\n",
      "Node:  Model Brain Activity\n",
      "\t [D] Model Brain Activity is A computational model of any measurable electrical or chemical activity in the real brain..\n",
      "\t [R] Model Brain Activity is a Entity.\n",
      "\t [R] Model Brain Activity is a model of Biological Brain Activity.\n",
      "\n",
      "Node:  Reconstructing Model Brain\n",
      "\t [R] Reconstructing Model Brain is a Brain Modeling Activity.\n",
      "\n",
      "Node:  Model NGV Unit\n",
      "\t [R] Model NGV Unit is a Model Brain Component.\n",
      "\t [R] Model NGV Unit is a model of NGV Unit.\n",
      "\n",
      "Node:  Vein\n",
      "\t [D] Vein is Veins are blood vessels that carry blood towards the heart..\n",
      "\t [R] Vein is a Blood Vessel.\n",
      "\n",
      "Node:  Blood Vessel\n",
      "\t [D] Blood Vessel is A blood vessel is a component of the vascular system that carries blood throughout in the body.\n",
      "\t [R] Blood Vessel is a Brain Vascular Component.\n",
      "\n",
      "Node:  Subject\n",
      "\t [R] Subject is a Entity.\n",
      "\t [R] Subject species Species.\n",
      "\n",
      "Node:  Testosterone\n",
      "\t [R] Testosterone is a Hormone.\n",
      "\n",
      "Node:  Excitatory Neuron\n",
      "\t [D] Excitatory Neuron is A neuron that releases excitatory neurotransmitters. These types of neurotransmitters have excitatory effects on the neuron, meaning they increase the likelihood that the neuron will fire an action potential. Some of the major excitatory neurotransmitters include epinephrine and norepinephrine..\n",
      "\t [S] Excitatory Neuron is also know as Excitatory Neuron.\n",
      "\t [S] Excitatory Neuron is referred to as Excitatory Neuron.\n",
      "\t [R] Excitatory Neuron is a Neurotransmitter Type.\n",
      "\n",
      "Node:  Organization\n",
      "\t [D] Organization is An organization is a social or legal institution such as a company, society, etc..\n",
      "\t [R] Organization is a Brain Modeling Agent.\n",
      "\n",
      "Node:  Model Brain Volume Building Workflow\n",
      "\t [R] Model Brain Volume Building Workflow is a Model Building Workflow.\n",
      "\t [R] Model Brain Volume Building Workflow generates Model Brain Volume.\n",
      "\t [R] Model Brain Volume Building Workflow uses Brain Volume Blueprint.\n",
      "\t [R] Model Brain Volume Building Workflow uses Model Neuron.\n",
      "\n",
      "Node:  Metabolite\n",
      "\t [R] Metabolite is a Molecule.\n",
      "\n",
      "Node:  Molecule\n",
      "\t [D] Molecule is A molecule is an electrically neutral group of two or more atoms held together by chemical bonds..\n",
      "\t [R] Molecule is a Biological Brain Component.\n",
      "\n",
      "Node:  Model Brain Volume Parameter\n",
      "\t [R] Model Brain Volume Parameter is a Model Brain Parameter.\n",
      "\n",
      "Node:  Molecular Compartment Feature\n",
      "\t [R] Molecular Compartment Feature is a Brain Component Feature.\n",
      "\n",
      "Node:  Model Brain Parameter Constraint\n",
      "\t [R] Model Brain Parameter Constraint is a Entity.\n",
      "\t [R] Model Brain Parameter Constraint constraints Model Brain Parameter.\n",
      "\n",
      "Node:  Neuron ME Type\n",
      "\t [S] Neuron ME Type is also know as METype.\n",
      "\t [S] Neuron ME Type is referred to as METype.\n",
      "\t [R] Neuron ME Type is a Neuron Type.\n",
      "\n",
      "Node:  Gene\n",
      "\t [D] Gene is In biology, a gene is a basic unit of heredity and a sequence of nucleotides in DNA or RNA that encodes the synthesis of a gene product, either RNA or protein..\n",
      "\t [R] Gene is a Biological Brain Component.\n",
      "\n",
      "Node:  Model Subcellular Part\n",
      "\t [R] Model Subcellular Part is a Model Brain Component.\n",
      "\t [R] Model Subcellular Part is a model of Subcellular Part.\n",
      "\n",
      "Node:  Sodium Channel\n",
      "\t [R] Sodium Channel is a Ion Channel.\n",
      "\t [R] Sodium Channel associatedIon Na.\n",
      "\n",
      "Node:  Na\n",
      "\t [S] Na is also know as Sodium.\n",
      "\t [S] Na is referred to as Sodium.\n",
      "\t [R] Na is a Ion.\n",
      "\n",
      "Node:  Model Cytosol Metabolism\n",
      "\t [R] Model Cytosol Metabolism is a Model Subcellular Part Metabolism.\n",
      "\t [R] Model Cytosol Metabolism is a model of Cytosol.\n",
      "\n",
      "Node:  Cytosol\n",
      "\t [R] Cytosol is a Subcellular Part.\n",
      "\n",
      "Node:  Inferred Data\n",
      "\t [R] Inferred Data is a Data Generation Modality.\n",
      "\n",
      "Node:  Glutamate\n",
      "\t [R] Glutamate is a Classical Neurotransmitter.\n",
      "\t [R] Glutamate is a Excitatory Transmitter.\n",
      "\n",
      "Node:  Cellular Component\n",
      "\t [D] Cellular Component is A location, relative to cellular compartments and structures, occupied by a macromolecular machine when it carries out a molecular function. There are two ways in which the gene ontology describes locations of gene products: (1) relative to cellular structures (e.g., cytoplasmic side of plasma membrane) or compartments (e.g., mitochondrion), and (2) the stable macromolecular complexes of which they are parts (e.g., the ribosome)..\n",
      "\t [S] Cellular Component is also know as cell or subcellular entity.\n",
      "\t [S] Cellular Component is referred to as cell or subcellular entity.\n",
      "\t [S] Cellular Component is also know as cellular component.\n",
      "\t [S] Cellular Component is referred to as cellular component.\n",
      "\t [R] Cellular Component http://www.w3.org/2002/07/owl#equivalentClass Subcellular Part.\n",
      "\n",
      "Node:  Brain Atlas Spatial Reference System\n",
      "\t [D] Brain Atlas Spatial Reference System is The Atlas Spatial Reference System is an entity which is the spatial reference system used as reference for a brain atlas. It can be described by information about the point of origin of the reference system, the orientation of the refernece system, the bounding box usde and the dimension unit of the brain atlas space..\n",
      "\t [R] Brain Atlas Spatial Reference System is a Entity.\n",
      "\n",
      "Node:  Circuit Connectivity Blueprint\n",
      "\t [R] Circuit Connectivity Blueprint is a Brain Component Blueprint.\n",
      "\n",
      "Node:  Person\n",
      "\t [D] Person is Person agents are people..\n",
      "\t [R] Person is a Brain Modeling Agent.\n",
      "\n",
      "Node:  Neocortical Thalamic System\n",
      "\t [D] Neocortical Thalamic System is The Neocortical Thalamic system is a brain system composed of two brain regions, the Neocortex and Thalamus with all the connections between them. The Neocortical Thalamic System includes all Brain Areas of the Neocortex and Thalamus and their intrinsic, afferent and efferent connections..\n",
      "\t [S] Neocortical Thalamic System is also know as NCT System.\n",
      "\t [S] Neocortical Thalamic System is referred to as NCT System.\n",
      "\t [R] Neocortical Thalamic System is a Brain System.\n",
      "\n",
      "Node:  Non-specific Channel\n",
      "\t [R] Non-specific Channel is a Ion Channel.\n",
      "\n",
      "Node:  Brain Location\n",
      "\t [R] Brain Location is a Entity.\n",
      "\t [R] Brain Location brainLayer Brain Layer.\n",
      "\t [R] Brain Location brainRegion Brain Region.\n",
      "\t [R] Brain Location coordinatesInAtlas Vector3D.\n",
      "\n",
      "Node:  Object Of Study\n",
      "\t [R] Object Of Study is a Entity.\n",
      "\n",
      "Node:  Neuron Morphological Type\n",
      "\t [D] Neuron Morphological Type is A classification scheme for a neuron related to its shape..\n",
      "\t [R] Neuron Morphological Type is a Neuron Type.\n",
      "\t [R] Neuron Morphological Type has feature Neuron Morphological Feature.\n",
      "\n",
      "Node:  Artery\n",
      "\t [D] Artery is An artery is a blood vessel that takes blood away from the heart..\n",
      "\t [R] Artery is a Blood Vessel.\n",
      "\n",
      "Node:  Data Release Workflow\n",
      "\t [R] Data Release Workflow is a Workflow.\n",
      "\n",
      "Node:  MEModel\n",
      "\t [R] MEModel is a Model Neuron.\n",
      "\t [R] MEModel is a model of Neuron ME Type.\n",
      "\n",
      "Node:  E3 Synapse Type\n",
      "\t [R] E3 Synapse Type is a Pseudo-Linear Synapse Type.\n",
      "\t [R] E3 Synapse Type is a Excitatory Synapse Type.\n",
      "\n",
      "Node:  Pseudo-Linear Synapse Type\n",
      "\t [R] Pseudo-Linear Synapse Type is a Synapse Plasticity Type.\n",
      "\n",
      "Node:  Excitatory Synapse Type\n",
      "\t [R] Excitatory Synapse Type is a Neurotransmitter Synapse Type.\n",
      "\n",
      "Node:  Rattus Norvegicus\n",
      "\t [S] Rattus Norvegicus is also know as Rat.\n",
      "\t [S] Rattus Norvegicus is referred to as Rat.\n",
      "\t [R] Rattus Norvegicus is a Species.\n",
      "\n",
      "Node:  Organelle\n",
      "\t [D] Organelle is Organelle is a Subcellular Component of cells with specific functions within a cell.\n",
      "\t [E] Examples of Organelle include Endoplasmic Reticulum, Mitochondria.\n",
      "\t [E] Endoplasmic Reticulum, Mitochondria is an example of Organelle.\n",
      "\t [E] Different examples of Organelle such as Endoplasmic Reticulum, Mitochondria.\n",
      "\t [R] Organelle is a Subcellular Part.\n",
      "\n",
      "Node:  Transient K+ Current\n",
      "\t [D] Transient K+ Current is The transient component of the potassium current.\n",
      "\t [S] Transient K+ Current is also know as Transient potassium current.\n",
      "\t [S] Transient K+ Current is referred to as Transient potassium current.\n",
      "\t [S] Transient K+ Current is also know as K_Tst.\n",
      "\t [S] Transient K+ Current is referred to as K_Tst.\n",
      "\t [R] Transient K+ Current is a Ionic Mechanism.\n",
      "\t [R] Transient K+ Current ion K.\n",
      "\n",
      "Node:  Ion Density Constraint Generation Workflow\n",
      "\t [R] Ion Density Constraint Generation Workflow is a EModel Parameter Constraint Generation Workflow.\n",
      "\t [R] Ion Density Constraint Generation Workflow generates Ion Channel Density Constraint.\n",
      "\n",
      "Node:  Protein\n",
      "\t [R] Protein is a Molecule.\n",
      "\n",
      "Node:  Offspring Size\n",
      "\t [D] Offspring Size is The size of offsprings in the genetic algorithm for optimization..\n",
      "\t [R] Offspring Size is a EModel Hyperparameter.\n",
      "\n",
      "Node:  Equation\n",
      "\t [D] Equation is In mathematics, an equation is a statement that asserts the equality of two expressions, which are connected by the equals sign \"=\"..\n",
      "\t [R] Equation is a Entity.\n",
      "\n",
      "Node:  Model Circuit Connectivity Building Workflow\n",
      "\t [R] Model Circuit Connectivity Building Workflow is a Model Building Workflow.\n",
      "\t [R] Model Circuit Connectivity Building Workflow generates Model Circuit Connectivity.\n",
      "\t [R] Model Circuit Connectivity Building Workflow uses Circuit Connectivity Blueprint.\n",
      "\t [R] Model Circuit Connectivity Building Workflow uses Model Synapse.\n",
      "\n",
      "Node:  Model Synapse\n",
      "\t [R] Model Synapse is a Model Brain Component.\n",
      "\t [R] Model Synapse is a model of Synapse.\n",
      "\n",
      "Node:  E2 Synapse Type\n",
      "\t [R] E2 Synapse Type is a Depressing Synapse Type.\n",
      "\t [R] E2 Synapse Type is a Excitatory Synapse Type.\n",
      "\n",
      "Node:  Biological Brain Activity\n",
      "\t [D] Biological Brain Activity is Brain Activity refers to any measurable electrical or chemical activity in the real brain..\n",
      "\t [R] Biological Brain Activity is a Entity.\n",
      "\n",
      "Node:  Rule\n",
      "\t [R] Rule is a Entity.\n",
      "\t [R] Rule isBasedOn Principle.\n",
      "\n",
      "Node:  Principle\n",
      "\t [R] Principle is a Entity.\n",
      "\n",
      "Node:  Neuron Metabolism Building Workflow\n",
      "\t [R] Neuron Metabolism Building Workflow is a Neuron Building Workflow.\n",
      "\t [R] Neuron Metabolism Building Workflow generates Model Neuron Metabolism.\n",
      "\t [R] Neuron Metabolism Building Workflow uses Model Biochemical Reaction.\n",
      "\t [R] Neuron Metabolism Building Workflow uses Model Subcellular Part Metabolism.\n",
      "\n",
      "Node:  WorkflowDefinition\n",
      "\t [R] WorkflowDefinition is a Entity.\n",
      "\n",
      "Node:  H Current\n",
      "\t [D] H Current is Hyperpolarization-activated current..\n",
      "\t [S] H Current is also know as Ih.\n",
      "\t [S] H Current is referred to as Ih.\n",
      "\t [R] H Current is a Ionic Mechanism.\n",
      "\n",
      "Node:  CA1 field of hippocampus\n",
      "\t [R] CA1 field of hippocampus is a Brain Area.\n",
      "\n",
      "Node:  Tau decay CaDynamics DC0\n",
      "\t [D] Tau decay CaDynamics DC0 is Time constant of calcium ion diffusion..\n",
      "\t [R] Tau decay CaDynamics DC0 is a Calcium Dynamics.\n",
      "\n",
      "Node:  Neuron Molecular Type\n",
      "\t [R] Neuron Molecular Type is a Neuron Type.\n",
      "\n",
      "Node:  Ca2+ Activated K+ Current\n",
      "\t [D] Ca2+ Activated K+ Current is Small-conductance calcium-activated potassium current..\n",
      "\t [S] Ca2+ Activated K+ Current is also know as SK_E2.\n",
      "\t [S] Ca2+ Activated K+ Current is referred to as SK_E2.\n",
      "\t [R] Ca2+ Activated K+ Current is a Ionic Mechanism.\n",
      "\t [R] Ca2+ Activated K+ Current ion K.\n",
      "\n",
      "Node:  VIP+\n",
      "\t [R] VIP+ is a Neuron Molecular Type.\n",
      "\n",
      "Node:  cAD\n",
      "\t [R] cAD is a Neuron Electrical Type.\n",
      "\n",
      "Node:  Data Generalization Rule\n",
      "\t [R] Data Generalization Rule is a Rule.\n",
      "\t [R] Data Generalization Rule premise Premise.\n",
      "\t [R] Data Generalization Rule searchQuery Query.\n",
      "\n",
      "Node:  Premise\n",
      "\t [R] Premise is a Entity.\n",
      "\n",
      "Node:  Circuit Connectivity Blueprint Building Workflow\n",
      "\t [R] Circuit Connectivity Blueprint Building Workflow is a Blueprint Building Workflow.\n",
      "\t [R] Circuit Connectivity Blueprint Building Workflow generates Circuit Connectivity Blueprint.\n",
      "\t [R] Circuit Connectivity Blueprint Building Workflow uses Bouton Density.\n",
      "\t [R] Circuit Connectivity Blueprint Building Workflow uses Model Brain Volume.\n",
      "\t [R] Circuit Connectivity Blueprint Building Workflow uses Number Of Synapses Per Connection.\n",
      "\n",
      "Node:  Bouton Density\n",
      "\t [R] Bouton Density is a Brain Volume Feature.\n",
      "\t [R] Bouton Density is a Model Connectivity Parameter.\n",
      "\t [R] Bouton Density brainLocation Brain Location.\n",
      "\t [R] Bouton Density cellType Brain Cell Type.\n",
      "\n",
      "Node:  Estrogen\n",
      "\t [R] Estrogen is a Hormone.\n",
      "\n",
      "Node:  PV+\n",
      "\t [R] PV+ is a Neuron Molecular Type.\n",
      "\n",
      "Node:  SST+\n",
      "\t [R] SST+ is a Neuron Molecular Type.\n",
      "\n",
      "Node:  Brain Atlas Release\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\t [D] Brain Atlas Release is The Atlas Release is an entity which describes the release of a brain atlas. An atlas release can be described by the release date and the template volume, parcellation volume, parcellation ontology and spatial reference system used..\n",
      "\t [R] Brain Atlas Release is a Entity.\n",
      "\n",
      "Node:  Oligodendrocyte\n",
      "\t [S] Oligodendrocyte is also know as Oligodendrocyte.\n",
      "\t [S] Oligodendrocyte is referred to as Oligodendrocyte.\n",
      "\t [R] Oligodendrocyte is a Glia.\n",
      "\n",
      "Node:  Facilitating Synapse Type\n",
      "\t [R] Facilitating Synapse Type is a Synapse Plasticity Type.\n",
      "\n",
      "Node:  Persistent Na+ Current\n",
      "\t [D] Persistent Na+ Current is Persistent sodium current..\n",
      "\t [R] Persistent Na+ Current is a Ionic Mechanism.\n",
      "\t [R] Persistent Na+ Current ion Na.\n",
      "\n",
      "Node:  Pyramidal Neuron\n",
      "\t [D] Pyramidal Neuron is A neuron with a pyramidal shaped soma and two distinct dendritic trees. The basal dendrites emerge from the base and the apical dendrites from the apex of the pyramidal cell body..\n",
      "\t [S] Pyramidal Neuron is also know as Pyramidal cell.\n",
      "\t [S] Pyramidal Neuron is referred to as Pyramidal cell.\n",
      "\t [R] Pyramidal Neuron is a Neuron Morphological Type.\n",
      "\n",
      "Node:  Brain Layer\n",
      "\t [D] Brain Layer is A Brain Layer is a layer of a laminar region in the brain. Laminar regions include e.g. the Neocortex or the Hippocampus..\n",
      "\t [E] Examples of Brain Layer include CA1 stratum oriens is an example of a brain layer in the Hippocampus., Layer 1 is an example of a brain layer in the Neocortex..\n",
      "\t [E] CA1 stratum oriens is an example of a brain layer in the Hippocampus., Layer 1 is an example of a brain layer in the Neocortex. is an example of Brain Layer.\n",
      "\t [E] Different examples of Brain Layer such as CA1 stratum oriens is an example of a brain layer in the Hippocampus., Layer 1 is an example of a brain layer in the Neocortex..\n",
      "\t [R] Brain Layer is a Biological Brain Component.\n",
      "\t [R] Brain Layer is a Brain Volume.\n",
      "\t [R] Brain Layer is a part of Brain Area.\n",
      "\t [R] Brain Layer is a part of Brain Region.\n",
      "\n",
      "Node:  Synaptic Strength\n",
      "\t [R] Synaptic Strength is a Synapse Feature.\n",
      "\n",
      "Node:  Axon\n",
      "\t [R] Axon is a Neuron Part.\n",
      "\t [R] Axon is a part of Neuron.\n",
      "\n",
      "Node:  Endoplasmic Reticulum\n",
      "\t [R] Endoplasmic Reticulum is a Organelle.\n",
      "\n",
      "Node:  Neocortical Layer\n",
      "\t [D] Neocortical Layer is One of the layers of the neocortex..\n",
      "\t [S] Neocortical Layer is also know as layer of neocortex.\n",
      "\t [S] Neocortical Layer is referred to as layer of neocortex.\n",
      "\t [S] Neocortical Layer is also know as Neocortical Layer.\n",
      "\t [S] Neocortical Layer is referred to as Neocortical Layer.\n",
      "\t [R] Neocortical Layer is a Brain Layer.\n",
      "\n",
      "Node:  Number Of Synapses\n",
      "\t [R] Number Of Synapses is a Synaptic Connection Feature.\n",
      "\n",
      "Node:  I1 Synapse Type\n",
      "\t [R] I1 Synapse Type is a Facilitating Synapse Type.\n",
      "\t [R] I1 Synapse Type is a Inhibitory Synapse Type.\n",
      "\n",
      "Node:  Transient Na+ Current\n",
      "\t [D] Transient Na+ Current is Transient sodium current..\n",
      "\t [R] Transient Na+ Current is a Ionic Mechanism.\n",
      "\t [R] Transient Na+ Current ion Na.\n",
      "\n",
      "Node:  Experimental Activity\n",
      "\t [D] Experimental Activity is An experimental activity is any activity associated with generating data from a biological brain in a lab or a digital tissue in silico..\n",
      "\t [R] Experimental Activity is a Activity.\n",
      "\n",
      "Node:  EModel Parameter Optimization Batch Size\n",
      "\t [D] EModel Parameter Optimization Batch Size is The number of optimization seeds to run in parallel for each batch of optimization launched by the Luigi pipeline..\n",
      "\t [R] EModel Parameter Optimization Batch Size is a EModel Hyperparameter.\n",
      "\n",
      "Node:  Capillary\n",
      "\t [D] Capillary is A capillary is the smallest blood vessel in the body..\n",
      "\t [R] Capillary is a Blood Vessel.\n",
      "\n",
      "Node:  Microglia\n",
      "\t [S] Microglia is also know as Microglia.\n",
      "\t [S] Microglia is referred to as Microglia.\n",
      "\t [R] Microglia is a Glia.\n",
      "\n",
      "Node:  MO-CMA\n",
      "\t [D] MO-CMA is Multi-objective covariance matrix adaptation evolution strategy..\n",
      "\t [R] MO-CMA is a EModel Parameter Optimizer.\n",
      "\t [R] MO-CMA is a Evolutionary Algorithm.\n",
      "\n",
      "Node:  Electrical Series\n",
      "\t [R] Electrical Series is a Time Series.\n",
      "\n",
      "Node:  Section Regions\n",
      "\t [R] Section Regions is a Neurite Location Feature.\n",
      "\n",
      "Node:  Neuron Morphological Feature\n",
      "\t [R] Neuron Morphological Feature is a Neuron Feature.\n",
      "\n",
      "Node:  Data Provenance\n",
      "\t [D] Data Provenance is Data provenance refers to records of the inputs, entities, systems, and processes that influence data of interest, providing a historical record of the data and its origins..\n",
      "\n",
      "Node:  Max Generations\n",
      "\t [D] Max Generations is Maximum number of generation in the genetic algorithm for optimization..\n",
      "\t [R] Max Generations is a EModel Hyperparameter.\n",
      "\n",
      "Node:  EModel Parameter Optimization Stochasticity\n",
      "\t [D] EModel Parameter Optimization Stochasticity is A flag indicating whether stochasticity should be enabled for the channels that can be stochastic..\n",
      "\t [R] EModel Parameter Optimization Stochasticity is a EModel Hyperparameter.\n",
      "\n",
      "Node:  Lamp+\n",
      "\t [R] Lamp+ is a Neuron Molecular Type.\n",
      "\n",
      "Node:  Ca_LVAst\n",
      "\t [D] Ca_LVAst is Low-voltage-gaterd calcium current.\n",
      "\t [R] Ca_LVAst is a Ionic Mechanism.\n",
      "\t [R] Ca_LVAst ion Ca.\n",
      "\n",
      "Node:  Glia Type\n",
      "\t [R] Glia Type is a Brain Cell Type.\n",
      "\n",
      "Node:  PerisomaticCurrentFiringFeature\n",
      "\t [D] PerisomaticCurrentFiringFeature is Feature of perisomatic firing response to current steps.\n",
      "\t [R] PerisomaticCurrentFiringFeature is a Neuron Electrophysiological Feature.\n",
      "\t [R] PerisomaticCurrentFiringFeature compartment Soma.\n",
      "\n",
      "Node:  EModel Validation Workflow\n",
      "\t [R] EModel Validation Workflow is a Model Validation Workflow.\n",
      "\t [R] EModel Validation Workflow hasTarget Neuron Electrophysiological Feature.\n",
      "\t [R] EModel Validation Workflow uses Trace.\n",
      "\n",
      "Node:  Extracellular Part Building Workflow\n",
      "\t [R] Extracellular Part Building Workflow is a Model Building Workflow.\n",
      "\t [R] Extracellular Part Building Workflow generates Model Extracellular Part.\n",
      "\n",
      "Node:  E1 Synapse Type\n",
      "\t [R] E1 Synapse Type is a Facilitating Synapse Type.\n",
      "\t [R] E1 Synapse Type is a Excitatory Synapse Type.\n",
      "\n",
      "Node:  Electrophysiological Feature Extraction Workflow\n",
      "\t [R] Electrophysiological Feature Extraction Workflow is a Feature Extraction Workflow.\n",
      "\t [R] Electrophysiological Feature Extraction Workflow is a part of EModel Building Workflow.\n",
      "\t [R] Electrophysiological Feature Extraction Workflow generates Neuron Electrophysiological Feature.\n",
      "\t [R] Electrophysiological Feature Extraction Workflow uses Trace.\n",
      "\n",
      "Node:  Mus Musculus\n",
      "\t [S] Mus Musculus is also know as M. musculus.\n",
      "\t [S] Mus Musculus is referred to as M. musculus.\n",
      "\t [S] Mus Musculus is also know as Mouse.\n",
      "\t [S] Mus Musculus is referred to as Mouse.\n",
      "\t [S] Mus Musculus is also know as house mouse.\n",
      "\t [S] Mus Musculus is referred to as house mouse.\n",
      "\t [R] Mus Musculus is a Species.\n",
      "\n",
      "Node:  Extracellular Part Metabolism Building Workflow\n",
      "\t [R] Extracellular Part Metabolism Building Workflow is a Extracellular Part Building Workflow.\n",
      "\t [R] Extracellular Part Metabolism Building Workflow generates Model Extracellular Part Metabolism.\n",
      "\t [R] Extracellular Part Metabolism Building Workflow about Extracellular Part.\n",
      "\t [R] Extracellular Part Metabolism Building Workflow uses Model Biochemical Reaction.\n",
      "\n",
      "Node:  Model Reaction Simulation Workflow\n",
      "\t [R] Model Reaction Simulation Workflow is a Brain Component Simulation Workflow.\n",
      "\t [R] Model Reaction Simulation Workflow generates Molecular Concentration Series.\n",
      "\t [R] Model Reaction Simulation Workflow uses Model Biochemical Reaction.\n",
      "\n",
      "Node:  Experimental Data\n",
      "\t [R] Experimental Data is a Data Generation Modality.\n",
      "\n",
      "Node:  Neuron Transcriptome Type\n",
      "\t [R] Neuron Transcriptome Type is a Neuron Type.\n",
      "\t [R] Neuron Transcriptome Type expressionProfile Gene Expression Profile.\n",
      "\t [R] Neuron Transcriptome Type gene Gene.\n",
      "\n",
      "Node:  Gene Expression Profile\n",
      "\t [R] Gene Expression Profile is a Dataset.\n",
      "\n",
      "Node:  Reaction Kinetic Parameter\n",
      "\t [D] Reaction Kinetic Parameter is Kinetic parameter is a parameter in the rate equation of the chemical process (reaction, transport, diffusion)..\n",
      "\t [R] Reaction Kinetic Parameter is a Model Biochemical Reaction Parameter.\n",
      "\n",
      "Node:  Noradrenaline\n",
      "\t [D] Noradrenaline is An organic chemical in the catecholamine family that functions in the brain and body as both a hormone and neurotransmitter. The general function of norepinephrine is to mobilize the brain and body for action. Norepinephrine release is lowest during sleep, rises during wakefulness, and reaches much higher levels during situations of stress or danger, in the so-called fight-or-flight response. In the brain, norepinephrine increases arousal and alertness, promotes vigilance, enhances formation and retrieval of memory, and focuses attention; it also increases restlessness and anxiety. In the rest of the body, norepinephrine increases heart rate and blood pressure, triggers the release of glucose from energy stores, increases blood flow to skeletal muscle, reduces blood flow to the gastrointestinal system, and inhibits voiding of the bladder and gastrointestinal motility..\n",
      "\t [S] Noradrenaline is also know as NA.\n",
      "\t [S] Noradrenaline is referred to as NA.\n",
      "\t [S] Noradrenaline is also know as NE.\n",
      "\t [S] Noradrenaline is referred to as NE.\n",
      "\t [S] Noradrenaline is also know as Noradrenalin.\n",
      "\t [S] Noradrenaline is referred to as Noradrenalin.\n",
      "\t [S] Noradrenaline is also know as Norepinephrine.\n",
      "\t [S] Noradrenaline is referred to as Norepinephrine.\n",
      "\t [R] Noradrenaline is a Neuromodulator.\n",
      "\n",
      "Node:  Ionic Mechanism Building Workflow\n",
      "\t [R] Ionic Mechanism Building Workflow is a Model Building Workflow.\n",
      "\t [R] Ionic Mechanism Building Workflow generates Model Ionic Mechanism.\n",
      "\t [R] Ionic Mechanism Building Workflow about Ionic Mechanism.\n",
      "\n",
      "Node:  Apical Dendrite\n",
      "\t [R] Apical Dendrite is a Dendrite.\n",
      "\n",
      "Node:  regional part of brain\n",
      "\t [D] regional part of brain is Anatomical divisons of the brain according to one or more criteria, e.g. cytoarchitectural, gross anatomy. Parts may be contiguous in 3D or not, e.g., basal ganglia..\n",
      "\t [S] regional part of brain is also know as regional part of brain.\n",
      "\t [S] regional part of brain is referred to as regional part of brain.\n",
      "\t [R] regional part of brain http://www.w3.org/2002/07/owl#equivalentClass Brain Region.\n",
      "\n",
      "Node:  Quantitative Value\n",
      "\t [R] Quantitative Value is a Entity.\n",
      "\n",
      "Node:  Steady State Concentration Extraction Workflow\n",
      "\t [R] Steady State Concentration Extraction Workflow is a Feature Extraction Workflow.\n",
      "\t [R] Steady State Concentration Extraction Workflow is a part of Model Reaction Building Workflow.\n",
      "\t [R] Steady State Concentration Extraction Workflow generates Steady State Molecular Concentration.\n",
      "\t [R] Steady State Concentration Extraction Workflow uses Molecular Concentration Series.\n",
      "\n",
      "Node:  Neuron Projection Feature\n",
      "\t [D] Neuron Projection Feature is A feature associated with a neuron projection, a process extending from a neural cell, such as axons or dendrites (collectively called neurites)..\n",
      "\t [R] Neuron Projection Feature is a Neuron Feature.\n",
      "\n",
      "Node:  Modulatory Synapse Type\n",
      "\t [R] Modulatory Synapse Type is a Neurotransmitter Synapse Type.\n",
      "\n",
      "Node:  I3 Synapse Type\n",
      "\t [R] I3 Synapse Type is a Inhibitory Synapse Type.\n",
      "\t [R] I3 Synapse Type is a Pseudo-Linear Synapse Type.\n",
      "\n",
      "Node:  Interneuron\n",
      "\t [S] Interneuron is also know as Interneuron.\n",
      "\t [S] Interneuron is referred to as Interneuron.\n",
      "\t [R] Interneuron is a Neuron.\n",
      "\n",
      "Node:  Cl\n",
      "\t [S] Cl is also know as Chloride.\n",
      "\t [S] Cl is referred to as Chloride.\n",
      "\t [R] Cl is a Ion.\n",
      "\n",
      "Node:  SKv3_1\n",
      "\t [D] SKv3_1 is Small-conductance voltage-gated Kv3.1 potassium current.\n",
      "\t [R] SKv3_1 is a Ionic Mechanism.\n",
      "\t [R] SKv3_1 ion K.\n",
      "\n",
      "Node:  IC Selector-Based Ion Channel Density Constraint Generation Workflow\n",
      "\t [R] IC Selector-Based Ion Channel Density Constraint Generation Workflow is a Ion Density Constraint Generation Workflow.\n",
      "\t [R] IC Selector-Based Ion Channel Density Constraint Generation Workflow uses Ion Channel Density Constraint.\n",
      "\t [R] IC Selector-Based Ion Channel Density Constraint Generation Workflow generates Ion Channel Density Constraint.\n",
      "\n",
      "Node:  Dopamine\n",
      "\t [D] Dopamine is A neuromodulatory molecule that plays several important roles in cells. It is an organic chemical of the catecholamine and phenethylamine families. In the brain, dopamine functions as a neurotransmitter—a chemical released by neurons (nerve cells) to send signals to other nerve cells. Neurotransmitters are synthesized in specific regions of the brain, but affect many regions systemically. The brain includes several distinct dopamine pathways, one of which plays a major role in the motivational component of reward-motivated behavior. The anticipation of most types of rewards increases the level of dopamine in the brain, and many addictive drugs increase dopamine release or block its reuptake into neurons following release..\n",
      "\t [R] Dopamine is a Excitatory Transmitter.\n",
      "\t [R] Dopamine is a Neuromodulator.\n",
      "\n",
      "Node:  CA2 field of hippocampus\n",
      "\t [R] CA2 field of hippocampus is a Brain Area.\n",
      "\n",
      "Node:  Complex\n",
      "\t [D] Complex is A protein complex or multiprotein complex is a group of two or more associated polypeptide chains..\n",
      "\t [R] Complex is a Molecule.\n",
      "\n",
      "Node:  Default Ion Channel Density Constraint Generation Workflow\n",
      "\t [R] Default Ion Channel Density Constraint Generation Workflow is a Ion Density Constraint Generation Workflow.\n",
      "\t [R] Default Ion Channel Density Constraint Generation Workflow generates Ion Channel Density Constraint.\n",
      "\n",
      "Node:  Model Synapse Building Workflow\n",
      "\t [R] Model Synapse Building Workflow is a Model Building Workflow.\n",
      "\t [R] Model Synapse Building Workflow generates Model Synapse.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for n in graph.nodes():\n",
    "    print(\"Node: \", n)\n",
    "    label = graph.nodes[n][\"label\"]\n",
    "    definition = graph.nodes[n][\"definition\"]\n",
    "    \n",
    "    record = {\"text\": \"\", \"spans\": []}\n",
    "    \n",
    "    if not is_nan(definition):\n",
    "        text = f\"{label} is {definition}.\"\n",
    "        print(\"\\t [D]\", text)\n",
    "        add_annotation(record, label, text, 0, len(label))\n",
    "    \n",
    "    synonyms = []\n",
    "\n",
    "    synonym = graph.nodes[n][\"synonym\"]\n",
    "    if not is_nan(synonym):\n",
    "        if isinstance(synonym, list):\n",
    "            synonyms += synonym\n",
    "        else:\n",
    "            synonyms.append(synonym)\n",
    "    \n",
    "    altLabel = graph.nodes[n][\"altLabel\"]\n",
    "    if not is_nan(altLabel):\n",
    "        if isinstance(altLabel, list):\n",
    "            synonyms += altLabel\n",
    "        else:\n",
    "            synonyms.append(altLabel)\n",
    "            \n",
    "    prefLabel = graph.nodes[n][\"prefLabel\"]\n",
    "    if not is_nan(prefLabel):\n",
    "        if isinstance(prefLabel, list):\n",
    "            synonyms += prefLabel\n",
    "        else:\n",
    "            synonyms.append(prefLabel)\n",
    "        \n",
    "    for synonym in synonyms:\n",
    "        text = f\"{label} is also know as {synonym}.\"\n",
    "        print(\"\\t [S]\", text)\n",
    "        add_annotation(record, label, text, 0, len(label))\n",
    "        \n",
    "        text = f\"{label} is referred to as {synonym}.\"\n",
    "        print(\"\\t [S]\", text)\n",
    "        add_annotation(record, label, text, 0, len(label))\n",
    "    \n",
    "    example = graph.nodes[n][\"example\"]\n",
    "    if not is_nan(example):\n",
    "        if isinstance(example, list):\n",
    "            example = \", \".join(example)\n",
    "        \n",
    "        text = f\"Examples of {label} include {example}.\"\n",
    "        print(\"\\t [E]\", text)\n",
    "        add_annotation(record, label, text, 12, 12 + len(label))\n",
    "        \n",
    "        text = f\"{example} is an example of {label}.\"\n",
    "        print(\"\\t [E]\", text)\n",
    "        local_start = len(example) + 18\n",
    "        add_annotation(record, label, text, local_start, local_start + len(label))\n",
    "    \n",
    "        text = f\"Different examples of {label} such as {example}.\"\n",
    "        print(\"\\t [E]\", text)\n",
    "        local_start = 22\n",
    "        add_annotation(record, label, text, local_start, local_start + len(label))\n",
    "        \n",
    "    for _, neighbor in graph.out_edges(n):\n",
    "        for r_type in graph.edges[n, neighbor][\"@type\"]:\n",
    "            if r_type in COMMON_EDGE_ALIASES:\n",
    "                r_type = COMMON_EDGE_ALIASES[r_type]\n",
    "            text = f\"{label} {r_type} {neighbor}.\"\n",
    "            print(\"\\t [R]\", text)\n",
    "            add_annotation(record, label, text, 0, len(label))\n",
    "            \n",
    "    synthetic_text.append(record)\n",
    "        \n",
    "    print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"../data/synthetic_texts.json\", \"w\") as f:\n",
    "    json.dump(synthetic_text, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<action> _register_one\n",
      "<succeeded> True\n"
     ]
    }
   ],
   "source": [
    "dataset = Dataset(\n",
    "    forge,\n",
    "    name=\"Synthetic Sentences from BMO\",\n",
    "    description=\"This dataset contains a set of annotated synthetic texts generated from the BMO.\")\n",
    "dataset.add_distribution(\n",
    "    \"../data/synthetic_texts.json\",\n",
    "    content_type=\"application/json\")\n",
    "forge.register(dataset)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2. Use Wikipedia to extract some more text"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 200,
   "metadata": {},
   "outputs": [],
   "source": [
    "MANUAL_MAP = {\n",
    "    \"EModel\": \"https://en.wikipedia.org/wiki/Biological_neuron_model\",\n",
    "    \"GABA\": \"https://en.wikipedia.org/wiki/GABA_receptor\",\n",
    "    \"Whole Brain\": \"https://en.wikipedia.org/wiki/Brain\",\n",
    "    \"K\": \"https://en.wikipedia.org/wiki/Potassium\",\n",
    "    \"Na\": \"https://en.wikipedia.org/wiki/Sodium\",\n",
    "    \"Ca\": \"https://en.wikipedia.org/wiki/Calcium\"\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [],
   "source": [
    "TERMS_TO_IGNORE = {\n",
    "    \"Kv11\",\n",
    "    \"Interstitial Space\",\n",
    "    \"Brain Area\",\n",
    "    \"IBEA\",\n",
    "    \"NGV Unit\",\n",
    "    \"MO-CMA\",\n",
    "    \"SO-CMA\",\n",
    "    \"Model Cell\",\n",
    "    \"Quantitative Value\",\n",
    "    \"Vector3D\",\n",
    "    \"Trace\",\n",
    "    \"Subject\",\n",
    "    \"Brain Layer\",\n",
    "    \"Model Glia\",\n",
    "    \"Lamp+\",\n",
    "    \"SST+\",\n",
    "    \"VIP+\",\n",
    "    \"cAD\",\n",
    "    \"PV+\",\n",
    "    \"MEModel\",\n",
    "    \"Neuron Part\"\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "WIKI_SEARCH_MAPPING = {}\n",
    "for node in graph.nodes():\n",
    "    if node not in TERMS_TO_IGNORE and node not in MANUAL_MAP:\n",
    "        res = search_wiki(node.lower())\n",
    "        if res is not None:\n",
    "            WIKI_SEARCH_MAPPING[node] = res[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [],
   "source": [
    "ALL_MAPPED_ENTITIES = {**WIKI_SEARCH_MAPPING, **MANUAL_MAP}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create a reverse mapping of wiki pages to BMO terms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {},
   "outputs": [],
   "source": [
    "wiki_to_bmo = {\n",
    "    v: f\"https://bbp.epfl.ch/ontologies/core/bmo/{k.replace(' ', '')}\"\n",
    "    for k, v in ALL_MAPPED_ENTITIES.items()\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2.2. Get Wiki texts for mapped articles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 218,
   "metadata": {},
   "outputs": [],
   "source": [
    "corpus = []\n",
    "for k, v in ALL_MAPPED_ENTITIES.items():\n",
    "    corpus.append(extract_full_wiki_text(v))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 219,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Random extracted text:\n",
      "\n",
      "A hormone (from the Greek participle ὁρμῶν, \"setting in motion\") is any member of a class of signaling molecules in multicellular organisms, that are transported by intricate biological processes to distant organs to regulate physiology and behavior. Hormones are required for the correct development of animals, plants and fungi. The lax definition of a hormone (as a signalling molecule that acts distant from its site of production) means that many different classes of molecule can be defined as hormones. Among the substances that can be considered hormones, are eicosanoids (e.g. prostaglandins and thromboxanes), steroids (e.g. oestrogen and brassinosteroid), amino acid derivatives (e.g. epinephrine and auxin), protein / peptides (e.g. insulin and CLE peptides) and gases (e.g ethylene and nitric oxide). Hormones are used to communicate between organs and tissues. In vertebrates, hormones are responsible for the regulation of many physiological processes and behavioral activities such as digestion, metabolism, respiration, sensory perception, sleep, excretion, lactation, stress induction, growth and development, movement, reproduction, and mood manipulation. In plants, hormones modulate almost all aspects of development, from germination to senescence. Hormones affect distant cells by binding to specific receptor proteins in the target cell, resulting in a change in cell function. When a hormone binds to the receptor, it results in the activation of a signal transduction pathway that typically activates gene transcription, resulting in increased expression of target proteins. Hormones can also act in rapid, non-genomic pathways that can be synergistic with genomic effects. Water-soluble hormones (such as peptides and amines) generally act on the surface of target cells via second messengers. Lipid soluble hormones, (such as steroids) generally pass through the plasma membranes of target cells (both cytoplasmic and nuclear) to act within their nuclei. A notable exception to this are brassinosteroids  (Brassinosteroids are a class of polyhydroxysteroids that have been recognized as a sixth class of plant hormones and may have utility as an anticancer drug for endocrine-responsive cancers to induce apoptosis and inhibit growth in plants) which despite being lipid soluble, still bind to their receptor at the cell surface. In vertebrates, endocrine glands are specialized organs that secrete hormones into the endocrine signaling system. Hormone secretion occurs in response to specific biochemical signals and is often subject to negative feedback regulation. For instance, high blood sugar (serum glucose concentration) promotes insulin synthesis. Insulin then acts to reduce glucose levels and maintain homeostasis, leading to reduced insulin levels.  Upon secretion, water soluble hormones are readily transported through the circulatory system. Lipid-soluble hormones must bond to carrier plasma glycoproteins (e.g., thyroxine-binding globulin (TBG)) to form ligand-protein complexes. Completely active hormones can be released into the bloodstream (as seen in insulin and growth hormones), but some travel as prohormones that must be activated in specific cells through a series of activation steps that are commonly highly regulated. The endocrine system secretes hormones directly into the bloodstream, typically via fenestrated capillaries, whereas the exocrine system secretes its hormones indirectly using ducts. Hormones with paracrine function diffuse through the interstitial spaces to nearby target tissue. Plants lack specialized organs for the secretion of hormones, although there is spatial distribution of hormone production. For example, the hormone auxin is produced mainly at the tips of young leaves and in the shoot apical meristem. The lack of specialised glands means that the main site of hormone production can change throughout the life of a plant, and the site of production is dependent on the plant's age and environment. Hormonal signaling involves the following steps: Hormone producing cells are found in the endocrine glands, such as the thyroid gland, ovaries, and testes. Exocytosis and other methods of membrane transport are used to secrete hormones when the endocrine glands are signaled. The hierarchical model is an oversimplification of the hormonal signaling process. Cellular recipients of a particular hormonal signal may be one of several cell types that reside within a number of different tissues, as is the case for insulin, which triggers a diverse range of systemic physiological effects. Different tissue types may also respond differently to the same hormonal signal. Arnold Adolph Berthold was a German physiologist and zoologist, who, in 1849, had a question about the function of the testes. He noticed in castrated roosters that they did not have the same sexual behaviors as roosters with their testes intact. He decided to run an experiment on male roosters to examine this phenomenon. He kept a group of roosters with their testes intact, and saw that they had normal sized wattles and combs (secondary sexual organs), a normal crow, and normal sexual and aggressive behaviors. He also had a group with their testes surgically removed, and noticed that their secondary sexual organs were decreased in size, had a weak crow, did not have sexual attraction towards females, and were not aggressive. He realized that this organ was essential for these behaviors, but he did not know how. To test this further, he removed one testis and placed it in the abdominal cavity. The roosters acted and had normal physical anatomy. He was able to see that location of the testes does not matter. He then wanted to see if it was a genetic factor that was involved in the testes that provided these functions. He transplanted a testis from another rooster to a rooster with one testis removed, and saw that they had normal behavior and physical anatomy as well. Berthold determined that the location or genetic factors of the testes do not matter in relation to sexual organs and behaviors, but that some chemical in the testes being secreted is causing this phenomenon. It was later identified that this factor was the hormone testosterone. Although known primarily for his work on the Theory of Evolution, Charles Darwin was also keenly interested in plants. Through the 1870s, he and his son Francis studied the movement of plants towards light. They were able to show that light is perceived at the tip of a young stem (the coleoptile), whereas the bending occurs lower down the stem. They proposed that a 'transmissible substance' communicated the direction of light from the tip down to the stem. The idea of a 'transmissible substance' was initially dismissed by other plant biologists, but their work later led to the discovery of the first plant hormone. In the 1920s Dutch scientist Frits Warmolt Went and Russian scientist Nikolai Cholodny (working independently of each other) conclusively showed that asymmetric accumulation of a growth hormone was responsible for this bending. In 1933 this hormone was finally isolated by Kögl, Haagen-Smit and Erxleben and given the name 'auxin'. William Bayliss and Ernest Starling, a physiologist and biologist, respectively, wanted to see if the nervous system had an impact on the digestive system. They knew that the pancreas was involved in the secretion of digestive fluids after the passage of food from the stomach to the intestines, which they believed to be due to the nervous system. They cut the nerves to the pancreas in an animal model and discovered that it was not nerve impulses that controlled secretion from the pancreas. It was determined that a factor secreted from the intestines into the bloodstream was stimulating the pancreas to secrete digestive fluids. This factor was named secretin: a hormone, although the term hormone was not coined until 1905 by Starling. Hormonal effects are dependent on where they are released, as they can be released in different manners. Not all hormones are released from a cell and into the blood until it binds to a receptor on a target.  The major types of hormone signaling are: As hormones are defined functionally, not structurally, they may have diverse chemical structures. Hormones occur in multicellular organisms (plants, animals, fungi, brown algae, and red algae). These compounds occur also in unicellular organisms, and may act as signaling molecules however there is no agreement that these molecules can be called hormones. Derivatines Compared with vertebrates, insects and crustaceans possess a number of structurally unusual hormones such as the juvenile hormone, a sesquiterpenoid. Examples include abscisic acid, auxin, cytokinin, ethylene, and gibberellin.  Most hormones initiate a cellular response by initially binding to either cell membrane associated or intracellular receptors. A cell may have several different receptor types that recognize the same hormone but activate different signal transduction pathways, or a cell may have several different receptors that recognize different hormones and activate the same biochemical pathway. Receptors for most peptide as well as many eicosanoid hormones are embedded in the plasma membrane at the surface of the cell and the majority of these receptors belong to the G protein-coupled receptor (GPCR) class of seven alpha helix transmembrane proteins. The interaction of hormone and receptor typically triggers a cascade of secondary effects within the cytoplasm of the cell, described as signal transduction, often involving phosphorylation or dephosphorylation of various other cytoplasmic proteins, changes in ion channel permeability, or increased concentrations of intracellular molecules that may act as secondary messengers (e.g., cyclic AMP). Some protein hormones also interact with intracellular receptors located in the cytoplasm or nucleus by an intracrine mechanism. For steroid or thyroid hormones, their receptors are located inside the cell within the cytoplasm of the target cell. These receptors belong to the nuclear receptor family of ligand-activated transcription factors. To bind their receptors, these hormones must first cross the cell membrane. They can do so because they are lipid-soluble. The combined hormone-receptor complex then moves across the nuclear membrane into the nucleus of the cell, where it binds to specific DNA sequences, regulating the expression of certain genes, and thereby increasing the levels of the proteins encoded by these genes. However, it has been shown that not all steroid receptors are located inside the cell. Some are associated with the plasma membrane. Hormones have the following effects on the body: A hormone may also regulate the production and release of other hormones. Hormone signals control the internal environment of the body through homeostasis. The rate of hormone biosynthesis and secretion is often regulated by a homeostatic negative feedback control mechanism. Such a mechanism depends on factors that influence the metabolism and excretion of hormones. Thus, higher hormone concentration alone cannot trigger the negative feedback mechanism. Negative feedback must be triggered by overproduction of an \"effect\" of the hormone. Hormone secretion can be stimulated and inhibited by: One special group of hormones is the tropic hormones that stimulate the hormone production of other endocrine glands. For example, thyroid-stimulating hormone (TSH) causes growth and increased activity of another endocrine gland, the thyroid, which increases output of thyroid hormones. To release active hormones quickly into the circulation, hormone biosynthetic cells may produce and store biologically inactive hormones in the form of pre- or prohormones. These can then be quickly converted into their active hormone form in response to a particular stimulus. Eicosanoids are considered to act as local hormones. They are considered to be \"local\" because they possess specific effects on target cells close to their site of formation. They also have a rapid degradation cycle, making sure they do not reach distant sites within the body. Hormones are also regulated by receptor agonists. Hormones are ligands, which are any kinds of molecules that produce a signal by binding to a receptor site on a protein. Hormone effects can be inhibited, thus regulated, by competing ligands that bind to the same target receptor as the hormone in question. When a competing ligand is bound to the receptor site, the hormone is unable to bind to that site and is unable to elicit a response from the target cell. These competing ligands are called antagonists of the hormone. Many hormones and their structural and functional analogs are used as medication. The most commonly prescribed hormones are estrogens and progestogens (as methods of hormonal contraception and as HRT), thyroxine (as levothyroxine, for hypothyroidism) and steroids (for autoimmune diseases and several respiratory disorders). Insulin is used by many diabetics. Local preparations for use in otolaryngology often contain pharmacologic equivalents of adrenaline, while steroid and vitamin D creams are used extensively in dermatological practice. A \"pharmacologic dose\" or \"supraphysiological dose\" of a hormone is a medical usage referring to an amount of a hormone far greater than naturally occurs in a healthy body. The effects of pharmacologic doses of hormones may be different from responses to naturally occurring amounts and may be therapeutically useful, though not without potentially adverse side effects. An example is the ability of pharmacologic doses of glucocorticoids to suppress inflammation. At the neurological level, behavior can be inferred based on hormone concentration, which in turn are influenced by hormone-release patterns; the numbers and locations of hormone receptors; and the efficiency of hormone receptors for those involved in gene transcription. Hormone concentration does not incite behavior, as that would undermine other external stimuli; however, it influences the system by increasing the probability of a certain event to occur. Not only can hormones influence behavior, but also behavior and the environment can influence hormone concentration. Thus, a feedback loop is formed, meaning behavior can affect hormone concentration, which in turn can affect behavior, which in turn can affect hormone concentration, and so on. For example, hormone-behavior feedback loops are essential in providing constancy to episodic hormone secretion, as the behaviors affected by episodically secreted hormones directly prevent the continuous release of said hormones. Three broad stages of reasoning may be used to determine if a specific hormone-behavior interaction is present within a system: There are various clear distinctions between hormones and neurotransmitters: Neurohormones are a type of hormone that share a commonality with neurotransmitters. They are produced by endocrine cells that receive input from neurons, or neuroendocrine cells. Both classic hormones and neurohormones are secreted by endocrine tissue; however, neurohormones are the result of a combination between endocrine reflexes and neural reflexes, creating a neuroendocrine pathway. While endocrine pathways produce chemical signals in the form of hormones, the neuroendocrine pathway involves the electrical signals of neurons. In this pathway, the result of the electrical signal produced by a neuron is the release of a chemical, which is the neurohormone. Finally, like a classic hormone, the neurohormone is released into the bloodstream to reach its target. Hormone transport and the involvement of binding proteins is an essential aspect when considering the function of hormones. The formation of a complex with a binding protein has several benefits: the effective half-life of the bound hormone is increased, and a reservoir of bound hormones is created, which evens the variations in concentration of unbound hormones (bound hormones will replace the unbound hormones when these are eliminated). An example of the usage of hormone-binding proteins is in the thyroxine-binding protein which carries up to 80% of all thyroxine in the body, a crucial element in regulating the meta\n"
     ]
    }
   ],
   "source": [
    "print(\"Random extracted text:\\n\")\n",
    "index = random.randint(0, len(corpus))\n",
    "print(corpus[index])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Extract 'related' articles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "unique_articles = set()\n",
    "for k in wiki_to_bmo:\n",
    "    unique_articles.add(k)\n",
    "    data, incoming_links = query_dbpedia(k)\n",
    "    print(k)\n",
    "#     print(\"In docs: \", len(incoming_links))\n",
    "    out_links = data.get(\"http://dbpedia.org/ontology/wikiPageWikiLink\", [])\n",
    "    out_links = [\n",
    "        el[\"value\"]\n",
    "        for el in out_links\n",
    "    ]\n",
    "#     unique_articles.update(incoming_links)\n",
    "    unique_articles.update(out_links)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 221,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "If we want to extract related acticles, could retrieve articles:  5845\n"
     ]
    }
   ],
   "source": [
    "print(\"If we want to extract related acticles, could retrieve articles: \", len(unique_articles))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 222,
   "metadata": {},
   "outputs": [],
   "source": [
    "# for el in unique_articles:\n",
    "#     url_wiki = dbpedia_uri_to_wiki(el)\n",
    "#     corpus.append(extract_full_wiki_text(url_wiki))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2.3. Use an existing pre-trained ER model\n",
    "\n",
    "https://github.com/egerber/spaCy-entity-linker"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 223,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<spacy_entity_linker.EntityLinker.EntityLinker at 0x7f966bf999d0>"
      ]
     },
     "execution_count": 223,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# initialize language model\n",
    "nlp = spacy.load(\"en_core_web_lg\")\n",
    "\n",
    "# add pipeline (declared through entry_points in setup.py)\n",
    "nlp.add_pipe(\"entityLinker\", last=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 224,
   "metadata": {},
   "outputs": [],
   "source": [
    "wiki_texts = []"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "WIKIDATA_TO_WIKI = {}\n",
    "for text in corpus:\n",
    "    try:\n",
    "        doc = nlp(text)            \n",
    "        record = {\"text\": str(doc), \"spans\": []}\n",
    "        # returns all entities in the whole document\n",
    "        all_linked_entities = doc._.linkedEntities\n",
    "        for i, entity in enumerate(all_linked_entities):\n",
    "            wikidata_url = entity.url\n",
    "            if wikidata_url not in WIKIDATA_TO_WIKI:\n",
    "                wiki_url = wiki_url_from_wikidata(wikidata_url.split(\"/\")[-1], \"en\")\n",
    "                WIKIDATA_TO_WIKI[wikidata_url] = wiki_url\n",
    "            else:\n",
    "                wiki_url = WIKIDATA_TO_WIKI[wikidata_url]\n",
    "            if wiki_url in wiki_to_bmo:\n",
    "                span = {\n",
    "                    \"text\": entity.span.text,\n",
    "                    \"start\": entity.span.start_char,\n",
    "                    \"end\": entity.span.end_char,\n",
    "                    \"label\": wiki_to_bmo[wiki_url]\n",
    "                }\n",
    "                record[\"spans\"].append(span)\n",
    "            if i > 30 and len(record[\"spans\"]) == 0:\n",
    "                break\n",
    "        if len(record[\"spans\"]) > 0:\n",
    "            wiki_texts.append(record)\n",
    "        \n",
    "    except Exception as e:\n",
    "        pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 228,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"../data/wiki_texts.json\", \"w\") as f:\n",
    "    json.dump(wiki_texts, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<action> _register_one\n",
      "<succeeded> True\n"
     ]
    }
   ],
   "source": [
    "dataset = Dataset(\n",
    "    forge,\n",
    "    name=\"Wikipedia texts for BMO entities\",\n",
    "    description=\"This dataset contains a set of annotated Wikipedia texts mentioning various BMO entities.\")\n",
    "dataset.add_distribution(\n",
    "    \"../data/wiki_texts.json\",\n",
    "    content_type=\"application/json\")\n",
    "forge.register(dataset)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "bg",
   "language": "python",
   "name": "bg"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
